{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Pipeline\n",
    "\n",
    "This notebook consists of three sections. In section [Data Exploration](#explore), we read the data and provide an overview of each of the modalities. In section [Comparisons](#compare), we investigate the features that are shared across different modalities, and finally, in section [Pseudo Bulking](#aggregate), we aggregate the single cell data modalities to obtain pseudobulks. \n",
    "\n",
    "Section [Comparisons](#compare) and section [Pseudo Bulking](#aggregate) are independent of each other. However, running the first cell of each subsection of section [Data Exploration](#explore), in which the data is loaded and saved into a variable, is necessary for running different code snippets of this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sp\n",
    "import anndata as ad\n",
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/shakiba/Desktop/integraph/data/data\n"
     ]
    }
   ],
   "source": [
    "# Path to the data\n",
    "cwd = os.getcwd()\n",
    "data_path = os.path.abspath(os.path.join(cwd, \"../data/data\"))\n",
    "print(data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read(file_name, dtype, sheet = None):\n",
    "    path = os.path.join(data_path, file_name + \".\" + dtype)\n",
    "    data = None\n",
    "    if dtype == \"csv\":\n",
    "        data = sp.read_csv(path, delimiter=',')\n",
    "    elif dtype == \"tsv\":\n",
    "        data = sp.read_csv(path, delimiter='\\t', dtype = str)\n",
    "    elif dtype == \"h5ad\":\n",
    "        data = sp.read_h5ad(path)\n",
    "    elif dtype == \"xlsx\": \n",
    "        data = ad.read_excel(path, sheet, dtype='str')\n",
    "    elif dtype == \"txt\":\n",
    "        data = sp.read_text(path, delimiter= \"\t\")\n",
    "    else:\n",
    "        pass\n",
    "    return data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Exploration <a id='explore'></a>\n",
    "\n",
    "This snippet of the notebook contains code for 1) seperating ADT and scRNA, 2) extracting bulkRNA protein names, 3) generating an overview of the data contained in the modalities\n",
    "Every file is saved in current folder."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ADT and SC RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ADT:  View of AnnData object with n_obs × n_vars = 836148 × 192\n",
      "    obs: 'Annotation_cluster_id', 'Annotation_cluster_name', 'Annotation_minor_subset', 'Annotation_major_subset', 'Annotation_cell_type', 'GEX_region', 'QC_ngenes', 'QC_total_UMI', 'QC_pct_mitochondrial', 'QC_scrub_doublet_scores', 'TCR_chain_composition', 'TCR_clone_ID', 'TCR_clone_count', 'TCR_clone_proportion', 'TCR_contains_unproductive', 'TCR_doublet', 'TCR_chain_TRA', 'TCR_v_gene_TRA', 'TCR_d_gene_TRA', 'TCR_j_gene_TRA', 'TCR_c_gene_TRA', 'TCR_productive_TRA', 'TCR_cdr3_TRA', 'TCR_umis_TRA', 'TCR_chain_TRA2', 'TCR_v_gene_TRA2', 'TCR_d_gene_TRA2', 'TCR_j_gene_TRA2', 'TCR_c_gene_TRA2', 'TCR_productive_TRA2', 'TCR_cdr3_TRA2', 'TCR_umis_TRA2', 'TCR_chain_TRB', 'TCR_v_gene_TRB', 'TCR_d_gene_TRB', 'TCR_j_gene_TRB', 'TCR_c_gene_TRB', 'TCR_productive_TRB', 'TCR_chain_TRB2', 'TCR_v_gene_TRB2', 'TCR_d_gene_TRB2', 'TCR_j_gene_TRB2', 'TCR_c_gene_TRB2', 'TCR_productive_TRB2', 'TCR_cdr3_TRB2', 'TCR_umis_TRB2', 'BCR_umis_HC', 'BCR_contig_qc_HC', 'BCR_functionality_HC', 'BCR_v_call_HC', 'BCR_v_score_HC', 'BCR_j_call_HC', 'BCR_j_score_HC', 'BCR_junction_aa_HC', 'BCR_total_mut_HC', 'BCR_s_mut_HC', 'BCR_r_mut_HC', 'BCR_c_gene_HC', 'BCR_clone_per_replicate_HC', 'BCR_clone_global_HC', 'BCR_clonal_abundance_HC', 'BCR_locus_LC', 'BCR_umis_LC', 'BCR_contig_qc_LC', 'BCR_functionality_LC', 'BCR_v_call_LC', 'BCR_v_score_LC', 'BCR_j_call_LC', 'BCR_j_score_LC', 'BCR_junction_aa_LC', 'BCR_total_mut_LC', 'BCR_s_mut_LC', 'BCR_r_mut_LC', 'BCR_c_gene_LC', 'COMBAT_ID', 'scRNASeq_sample_ID', 'COMBAT_participant_timepoint_ID', 'Source', 'Age', 'Sex', 'Race', 'BMI', 'Hospitalstay', 'Death28', 'Institute', 'PreExistingHeartDisease', 'PreExistingLungDisease', 'PreExistingKidneyDisease', 'PreExistingDiabetes', 'PreExistingHypertension', 'PreExistingImmunocompromised', 'Smoking', 'Symptomatic', 'Requiredvasoactive', 'Respiratorysupport', 'SARSCoV2PCR', 'Outcome', 'TimeSinceOnset', 'Ethnicity', 'Tissue', 'DiseaseClassification', 'Pool_ID', 'Channel_ID'\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "    uns: 'Institute', 'ObjectCreateDate', 'Source_colors', 'Technology', 'genome_annotation_version'\n",
      "    obsm: 'X_umap', 'X_umap_source'\n",
      "    layers: 'raw' \n",
      "\n",
      "\n",
      "single cell RNA:  View of AnnData object with n_obs × n_vars = 836148 × 20615\n",
      "    obs: 'Annotation_cluster_id', 'Annotation_cluster_name', 'Annotation_minor_subset', 'Annotation_major_subset', 'Annotation_cell_type', 'GEX_region', 'QC_ngenes', 'QC_total_UMI', 'QC_pct_mitochondrial', 'QC_scrub_doublet_scores', 'TCR_chain_composition', 'TCR_clone_ID', 'TCR_clone_count', 'TCR_clone_proportion', 'TCR_contains_unproductive', 'TCR_doublet', 'TCR_chain_TRA', 'TCR_v_gene_TRA', 'TCR_d_gene_TRA', 'TCR_j_gene_TRA', 'TCR_c_gene_TRA', 'TCR_productive_TRA', 'TCR_cdr3_TRA', 'TCR_umis_TRA', 'TCR_chain_TRA2', 'TCR_v_gene_TRA2', 'TCR_d_gene_TRA2', 'TCR_j_gene_TRA2', 'TCR_c_gene_TRA2', 'TCR_productive_TRA2', 'TCR_cdr3_TRA2', 'TCR_umis_TRA2', 'TCR_chain_TRB', 'TCR_v_gene_TRB', 'TCR_d_gene_TRB', 'TCR_j_gene_TRB', 'TCR_c_gene_TRB', 'TCR_productive_TRB', 'TCR_chain_TRB2', 'TCR_v_gene_TRB2', 'TCR_d_gene_TRB2', 'TCR_j_gene_TRB2', 'TCR_c_gene_TRB2', 'TCR_productive_TRB2', 'TCR_cdr3_TRB2', 'TCR_umis_TRB2', 'BCR_umis_HC', 'BCR_contig_qc_HC', 'BCR_functionality_HC', 'BCR_v_call_HC', 'BCR_v_score_HC', 'BCR_j_call_HC', 'BCR_j_score_HC', 'BCR_junction_aa_HC', 'BCR_total_mut_HC', 'BCR_s_mut_HC', 'BCR_r_mut_HC', 'BCR_c_gene_HC', 'BCR_clone_per_replicate_HC', 'BCR_clone_global_HC', 'BCR_clonal_abundance_HC', 'BCR_locus_LC', 'BCR_umis_LC', 'BCR_contig_qc_LC', 'BCR_functionality_LC', 'BCR_v_call_LC', 'BCR_v_score_LC', 'BCR_j_call_LC', 'BCR_j_score_LC', 'BCR_junction_aa_LC', 'BCR_total_mut_LC', 'BCR_s_mut_LC', 'BCR_r_mut_LC', 'BCR_c_gene_LC', 'COMBAT_ID', 'scRNASeq_sample_ID', 'COMBAT_participant_timepoint_ID', 'Source', 'Age', 'Sex', 'Race', 'BMI', 'Hospitalstay', 'Death28', 'Institute', 'PreExistingHeartDisease', 'PreExistingLungDisease', 'PreExistingKidneyDisease', 'PreExistingDiabetes', 'PreExistingHypertension', 'PreExistingImmunocompromised', 'Smoking', 'Symptomatic', 'Requiredvasoactive', 'Respiratorysupport', 'SARSCoV2PCR', 'Outcome', 'TimeSinceOnset', 'Ethnicity', 'Tissue', 'DiseaseClassification', 'Pool_ID', 'Channel_ID'\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "    uns: 'Institute', 'ObjectCreateDate', 'Source_colors', 'Technology', 'genome_annotation_version'\n",
      "    obsm: 'X_umap', 'X_umap_source'\n",
      "    layers: 'raw' \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ADT_and_SC_RNA = read(\"COMBAT-CITESeq-DATA\", \"h5ad\")\n",
    "adt_names = []\n",
    "scRNA_names = []\n",
    "\n",
    "# Determine ADT feature names and RNA feature names\n",
    "for name in ADT_and_SC_RNA.var_names: \n",
    "    if name.startswith(\"AB_\"): # ADT data start with AB_ \n",
    "        adt_names.append(name)\n",
    "    else:\n",
    "        scRNA_names.append(name)\n",
    "\n",
    "# Extract ADT and scRNA \n",
    "ADT = ADT_and_SC_RNA[:, adt_names]\n",
    "scRNA = ADT_and_SC_RNA[:, scRNA_names]\n",
    "\n",
    "# Remove the prefix AB_ from ADT features \n",
    "adt_names = [name.split(\"AB_\")[1] for name in adt_names]\n",
    "\n",
    "print(\"ADT: \", ADT, \"\\n\\n\")\n",
    "print(\"single cell RNA: \", scRNA, \"\\n\\n\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ADT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ -67.67877 , 460.98032 ]\n"
     ]
    }
   ],
   "source": [
    "summary_ADT = {\"feature\" : adt_names,\n",
    "           \"min\" : ADT.X.toarray().min(axis = 0),\n",
    "           \"max\" : ADT.X.toarray().max(axis = 0),\n",
    "           \"mean\" : ADT.X.toarray().mean(axis = 0),\n",
    "           \"var\" : ADT.X.toarray().var(axis = 0)}\n",
    "\n",
    "df_adt= pd.DataFrame(summary_ADT)\n",
    "\n",
    "print(\"[\", ADT.X.toarray().min(), \",\", ADT.X.toarray().max(), \"]\") # Overall range of proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AB_CD80</th>\n",
       "      <th>AB_CD86</th>\n",
       "      <th>AB_CD274_B7_H1_PD_L1</th>\n",
       "      <th>AB_CD273_B7_DC_PD_L2</th>\n",
       "      <th>AB_CD275_B7_H2_ICOSL</th>\n",
       "      <th>AB_humanCD11b</th>\n",
       "      <th>AB_CD252_OX40L</th>\n",
       "      <th>AB_CD137L_4_1BBLigand</th>\n",
       "      <th>AB_CD155_PVR</th>\n",
       "      <th>AB_CD112_Nectin_2</th>\n",
       "      <th>...</th>\n",
       "      <th>AB_CD101_BB27</th>\n",
       "      <th>AB_CD360_IL_21R</th>\n",
       "      <th>AB_CD88_C5aR</th>\n",
       "      <th>AB_HLA_F</th>\n",
       "      <th>AB_NLRP2</th>\n",
       "      <th>AB_Podocalyxin</th>\n",
       "      <th>AB_CD224</th>\n",
       "      <th>AB_c_Met</th>\n",
       "      <th>AB_CD258_LIGHT</th>\n",
       "      <th>AB_DR3_TRAMP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAAGTGG-1-gPlexA1</th>\n",
       "      <td>1.987870</td>\n",
       "      <td>1.921781</td>\n",
       "      <td>2.613414</td>\n",
       "      <td>0.456505</td>\n",
       "      <td>1.482558</td>\n",
       "      <td>1.789405</td>\n",
       "      <td>1.206598</td>\n",
       "      <td>2.821688</td>\n",
       "      <td>1.885517</td>\n",
       "      <td>1.059094</td>\n",
       "      <td>...</td>\n",
       "      <td>0.840967</td>\n",
       "      <td>2.217810</td>\n",
       "      <td>1.220086</td>\n",
       "      <td>0.124530</td>\n",
       "      <td>1.719992</td>\n",
       "      <td>1.028112</td>\n",
       "      <td>1.729305</td>\n",
       "      <td>1.463706</td>\n",
       "      <td>1.785078</td>\n",
       "      <td>1.601881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCGGATCA-1-gPlexA1</th>\n",
       "      <td>-0.539351</td>\n",
       "      <td>0.442409</td>\n",
       "      <td>2.392834</td>\n",
       "      <td>1.047547</td>\n",
       "      <td>0.131874</td>\n",
       "      <td>1.147668</td>\n",
       "      <td>0.541517</td>\n",
       "      <td>1.990631</td>\n",
       "      <td>2.284331</td>\n",
       "      <td>1.762416</td>\n",
       "      <td>...</td>\n",
       "      <td>3.184458</td>\n",
       "      <td>2.731340</td>\n",
       "      <td>1.006678</td>\n",
       "      <td>-0.142191</td>\n",
       "      <td>1.202440</td>\n",
       "      <td>1.168217</td>\n",
       "      <td>3.295960</td>\n",
       "      <td>0.679976</td>\n",
       "      <td>2.942873</td>\n",
       "      <td>2.066820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGACATTA-1-gPlexA1</th>\n",
       "      <td>0.993282</td>\n",
       "      <td>1.441381</td>\n",
       "      <td>0.310766</td>\n",
       "      <td>-0.556409</td>\n",
       "      <td>1.708025</td>\n",
       "      <td>0.195210</td>\n",
       "      <td>0.375910</td>\n",
       "      <td>2.114272</td>\n",
       "      <td>-0.618993</td>\n",
       "      <td>-0.049037</td>\n",
       "      <td>...</td>\n",
       "      <td>0.937668</td>\n",
       "      <td>0.563752</td>\n",
       "      <td>0.735085</td>\n",
       "      <td>0.130218</td>\n",
       "      <td>-0.380042</td>\n",
       "      <td>0.243760</td>\n",
       "      <td>0.863445</td>\n",
       "      <td>0.484824</td>\n",
       "      <td>0.713824</td>\n",
       "      <td>1.316770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCGACAT-1-gPlexA1</th>\n",
       "      <td>0.838407</td>\n",
       "      <td>2.641940</td>\n",
       "      <td>0.344012</td>\n",
       "      <td>0.189955</td>\n",
       "      <td>1.021477</td>\n",
       "      <td>2.683172</td>\n",
       "      <td>1.320957</td>\n",
       "      <td>0.759884</td>\n",
       "      <td>1.355680</td>\n",
       "      <td>3.498369</td>\n",
       "      <td>...</td>\n",
       "      <td>0.947138</td>\n",
       "      <td>-0.744993</td>\n",
       "      <td>1.814052</td>\n",
       "      <td>1.673030</td>\n",
       "      <td>1.307825</td>\n",
       "      <td>1.210711</td>\n",
       "      <td>2.648582</td>\n",
       "      <td>0.606611</td>\n",
       "      <td>1.100374</td>\n",
       "      <td>-0.663722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGGAACGG-1-gPlexA1</th>\n",
       "      <td>1.172756</td>\n",
       "      <td>14.549344</td>\n",
       "      <td>-0.884014</td>\n",
       "      <td>1.349209</td>\n",
       "      <td>1.489393</td>\n",
       "      <td>4.734301</td>\n",
       "      <td>1.522198</td>\n",
       "      <td>0.871919</td>\n",
       "      <td>5.137852</td>\n",
       "      <td>4.039271</td>\n",
       "      <td>...</td>\n",
       "      <td>1.493470</td>\n",
       "      <td>2.687017</td>\n",
       "      <td>5.535050</td>\n",
       "      <td>1.429137</td>\n",
       "      <td>0.657836</td>\n",
       "      <td>2.042859</td>\n",
       "      <td>6.367860</td>\n",
       "      <td>1.316507</td>\n",
       "      <td>1.871877</td>\n",
       "      <td>1.279506</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 192 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             AB_CD80    AB_CD86  AB_CD274_B7_H1_PD_L1   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1  1.987870   1.921781              2.613414  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1 -0.539351   0.442409              2.392834   \n",
       "AAACCTGAGGACATTA-1-gPlexA1  0.993282   1.441381              0.310766   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1  0.838407   2.641940              0.344012   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1  1.172756  14.549344             -0.884014   \n",
       "\n",
       "                            AB_CD273_B7_DC_PD_L2  AB_CD275_B7_H2_ICOSL   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1              0.456505              1.482558  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1              1.047547              0.131874   \n",
       "AAACCTGAGGACATTA-1-gPlexA1             -0.556409              1.708025   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1              0.189955              1.021477   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1              1.349209              1.489393   \n",
       "\n",
       "                            AB_humanCD11b  AB_CD252_OX40L   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1       1.789405        1.206598  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1       1.147668        0.541517   \n",
       "AAACCTGAGGACATTA-1-gPlexA1       0.195210        0.375910   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1       2.683172        1.320957   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1       4.734301        1.522198   \n",
       "\n",
       "                            AB_CD137L_4_1BBLigand  AB_CD155_PVR   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1               2.821688      1.885517  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1               1.990631      2.284331   \n",
       "AAACCTGAGGACATTA-1-gPlexA1               2.114272     -0.618993   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1               0.759884      1.355680   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1               0.871919      5.137852   \n",
       "\n",
       "                            AB_CD112_Nectin_2  ...  AB_CD101_BB27   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1           1.059094  ...       0.840967  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1           1.762416  ...       3.184458   \n",
       "AAACCTGAGGACATTA-1-gPlexA1          -0.049037  ...       0.937668   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1           3.498369  ...       0.947138   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1           4.039271  ...       1.493470   \n",
       "\n",
       "                            AB_CD360_IL_21R  AB_CD88_C5aR  AB_HLA_F  AB_NLRP2   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1         2.217810      1.220086  0.124530  1.719992  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1         2.731340      1.006678 -0.142191  1.202440   \n",
       "AAACCTGAGGACATTA-1-gPlexA1         0.563752      0.735085  0.130218 -0.380042   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1        -0.744993      1.814052  1.673030  1.307825   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1         2.687017      5.535050  1.429137  0.657836   \n",
       "\n",
       "                            AB_Podocalyxin  AB_CD224  AB_c_Met   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1        1.028112  1.729305  1.463706  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1        1.168217  3.295960  0.679976   \n",
       "AAACCTGAGGACATTA-1-gPlexA1        0.243760  0.863445  0.484824   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1        1.210711  2.648582  0.606611   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1        2.042859  6.367860  1.316507   \n",
       "\n",
       "                            AB_CD258_LIGHT  AB_DR3_TRAMP  \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1        1.785078      1.601881  \n",
       "AAACCTGAGCGGATCA-1-gPlexA1        2.942873      2.066820  \n",
       "AAACCTGAGGACATTA-1-gPlexA1        0.713824      1.316770  \n",
       "AAACCTGAGGCGACAT-1-gPlexA1        1.100374     -0.663722  \n",
       "AAACCTGAGGGAACGG-1-gPlexA1        1.871877      1.279506  \n",
       "\n",
       "[5 rows x 192 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Content overview\n",
    "ADT.to_df().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save ADT information\n",
    "ADT.write_h5ad(\"adt.h5ad\")\n",
    "\n",
    "with open ('adt_feature_names.txt', 'w') as file:  \n",
    "    for name in adt_names:\n",
    "        file.write(name + \"\\n\") \n",
    "        \n",
    "df_adt.to_excel(\"summary_ADT.xlsx\") "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### scRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'matrix' object has no attribute 'toarray'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m summary_scRNA \u001b[39m=\u001b[39m {\u001b[39m\"\u001b[39m\u001b[39mfeature\u001b[39m\u001b[39m\"\u001b[39m : scRNA_names,\n\u001b[1;32m      2\u001b[0m            \u001b[39m\"\u001b[39m\u001b[39mmin\u001b[39m\u001b[39m\"\u001b[39m : scRNA\u001b[39m.\u001b[39mX\u001b[39m.\u001b[39mmin(axis \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m)\u001b[39m.\u001b[39mtoarray()[\u001b[39m0\u001b[39m],\n\u001b[1;32m      3\u001b[0m            \u001b[39m\"\u001b[39m\u001b[39mmax\u001b[39m\u001b[39m\"\u001b[39m : scRNA\u001b[39m.\u001b[39mX\u001b[39m.\u001b[39mmax(axis \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m)\u001b[39m.\u001b[39mtoarray()[\u001b[39m0\u001b[39m],\n\u001b[0;32m----> 4\u001b[0m            \u001b[39m\"\u001b[39m\u001b[39mmean\u001b[39m\u001b[39m\"\u001b[39m : scRNA\u001b[39m.\u001b[39;49mX\u001b[39m.\u001b[39;49mmean(axis \u001b[39m=\u001b[39;49m \u001b[39m0\u001b[39;49m)\u001b[39m.\u001b[39;49mtoarray()[\u001b[39m0\u001b[39m]}\n\u001b[1;32m      6\u001b[0m df_scRNA\u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(summary_scRNA)\n\u001b[1;32m      8\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39m[\u001b[39m\u001b[39m\"\u001b[39m, scRNA\u001b[39m.\u001b[39mX\u001b[39m.\u001b[39mmin(), \u001b[39m\"\u001b[39m\u001b[39m,\u001b[39m\u001b[39m\"\u001b[39m, scRNA\u001b[39m.\u001b[39mX\u001b[39m.\u001b[39mmax(), \u001b[39m\"\u001b[39m\u001b[39m]\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'matrix' object has no attribute 'toarray'"
     ]
    }
   ],
   "source": [
    "summary_scRNA = {\"feature\" : scRNA_names,\n",
    "           \"min\" : scRNA.X.min(axis = 0).toarray()[0],\n",
    "           \"max\" : scRNA.X.max(axis = 0).toarray()[0]}\n",
    "\n",
    "df_scRNA= pd.DataFrame(summary_scRNA)\n",
    "\n",
    "print(\"[\", scRNA.X.min(), \",\", scRNA.X.max(), \"]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OR4F5</th>\n",
       "      <th>OR4F29</th>\n",
       "      <th>OR4F16</th>\n",
       "      <th>SAMD11</th>\n",
       "      <th>NOC2L</th>\n",
       "      <th>KLHL17</th>\n",
       "      <th>PLEKHN1</th>\n",
       "      <th>PERM1</th>\n",
       "      <th>HES4</th>\n",
       "      <th>ISG15</th>\n",
       "      <th>...</th>\n",
       "      <th>AC007325.2</th>\n",
       "      <th>BX072566.1</th>\n",
       "      <th>AL354822.1</th>\n",
       "      <th>AC023491.2</th>\n",
       "      <th>AC004556.3</th>\n",
       "      <th>AC233755.2</th>\n",
       "      <th>AC233755.1</th>\n",
       "      <th>AC240274.1</th>\n",
       "      <th>AC213203.4</th>\n",
       "      <th>AC213203.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAAGTGG-1-gPlexA1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.553033</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCGGATCA-1-gPlexA1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.426129</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGACATTA-1-gPlexA1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCGACAT-1-gPlexA1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.800563</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.407496</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGGAACGG-1-gPlexA1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.445163</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.293625</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 20615 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            OR4F5  OR4F29  OR4F16  SAMD11     NOC2L  KLHL17   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1    0.0     0.0     0.0     0.0  1.553033     0.0  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1    0.0     0.0     0.0     0.0  0.000000     0.0   \n",
       "AAACCTGAGGACATTA-1-gPlexA1    0.0     0.0     0.0     0.0  0.000000     0.0   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1    0.0     0.0     0.0     0.0  1.800563     0.0   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1    0.0     0.0     0.0     0.0  1.445163     0.0   \n",
       "\n",
       "                            PLEKHN1  PERM1  HES4     ISG15  ...  AC007325.2   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1      0.0    0.0   0.0  0.000000  ...         0.0  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1      0.0    0.0   0.0  1.426129  ...         0.0   \n",
       "AAACCTGAGGACATTA-1-gPlexA1      0.0    0.0   0.0  0.000000  ...         0.0   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1      0.0    0.0   0.0  2.407496  ...         0.0   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1      0.0    0.0   0.0  3.293625  ...         0.0   \n",
       "\n",
       "                            BX072566.1  AL354822.1  AC023491.2  AC004556.3   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1         0.0         0.0         0.0         0.0  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1         0.0         0.0         0.0         0.0   \n",
       "AAACCTGAGGACATTA-1-gPlexA1         0.0         0.0         0.0         0.0   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1         0.0         0.0         0.0         0.0   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1         0.0         0.0         0.0         0.0   \n",
       "\n",
       "                            AC233755.2  AC233755.1  AC240274.1  AC213203.4   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1         0.0         0.0         0.0         0.0  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1         0.0         0.0         0.0         0.0   \n",
       "AAACCTGAGGACATTA-1-gPlexA1         0.0         0.0         0.0         0.0   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1         0.0         0.0         0.0         0.0   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1         0.0         0.0         0.0         0.0   \n",
       "\n",
       "                            AC213203.1  \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1         0.0  \n",
       "AAACCTGAGCGGATCA-1-gPlexA1         0.0  \n",
       "AAACCTGAGGACATTA-1-gPlexA1         0.0  \n",
       "AAACCTGAGGCGACAT-1-gPlexA1         0.0  \n",
       "AAACCTGAGGGAACGG-1-gPlexA1         0.0  \n",
       "\n",
       "[5 rows x 20615 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Data overview\n",
    "scRNA.to_df().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save scRNA information\n",
    "scRNA.write_h5ad(\"scRNA.h5ad\")\n",
    "\n",
    "with open ('scRNA_feature_names.txt', 'w') as file:  \n",
    "    for name in scRNA_names:\n",
    "        file.write(name + \"\\n\")  \n",
    "        \n",
    "df_scRNA.to_excel(\"summary_scRNA.xlsx\") "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### bulk RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bulk RNA:  AnnData object with n_obs × n_vars = 23063 × 143 \n",
      "\n",
      "\n",
      "bulk RNA:  AnnData object with n_obs × n_vars = 23063 × 4 \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/thesis/lib/python3.10/site-packages/anndata/_core/anndata.py:1828: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S00016-Ja001T-TRGa</th>\n",
       "      <th>S00020-Ja003T-TRGa</th>\n",
       "      <th>S00024-Ja003T-TRGa</th>\n",
       "      <th>S00027-Ja003T-TRGa</th>\n",
       "      <th>S00028-Ja001T-TRGa</th>\n",
       "      <th>S00030-Ja003T-TRGa</th>\n",
       "      <th>S00033-Ja001T-TRGa</th>\n",
       "      <th>S00033-Ja003T-TRGa</th>\n",
       "      <th>S00034-Ja005T-TRGa</th>\n",
       "      <th>S00033-Ja005T-TRGa</th>\n",
       "      <th>...</th>\n",
       "      <th>S00094-Ja005T-TRGa</th>\n",
       "      <th>S00095-Ja005T-TRGa</th>\n",
       "      <th>S00096-Ja005T-TRGa</th>\n",
       "      <th>S00097-Ja003T-TRGa</th>\n",
       "      <th>S00099-Ja005T-TRGa</th>\n",
       "      <th>S00104-Ja003T-TRGa</th>\n",
       "      <th>S00106-Ja003T-TRGa</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>membership</th>\n",
       "      <th>p.value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>0.358735</td>\n",
       "      <td>0.420317</td>\n",
       "      <td>0.212498</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.420037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044182</td>\n",
       "      <td>0.037569</td>\n",
       "      <td>0.266312</td>\n",
       "      <td>...</td>\n",
       "      <td>0.240967</td>\n",
       "      <td>0.269932</td>\n",
       "      <td>0.189985</td>\n",
       "      <td>0.285038</td>\n",
       "      <td>0.352373</td>\n",
       "      <td>0.581883</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.634714068658144</td>\n",
       "      <td>1.70946329554292e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>3.933231</td>\n",
       "      <td>4.017549</td>\n",
       "      <td>4.351443</td>\n",
       "      <td>4.102646</td>\n",
       "      <td>4.466349</td>\n",
       "      <td>3.650819</td>\n",
       "      <td>4.111353</td>\n",
       "      <td>4.330951</td>\n",
       "      <td>4.395805</td>\n",
       "      <td>4.293557</td>\n",
       "      <td>...</td>\n",
       "      <td>3.874222</td>\n",
       "      <td>3.827185</td>\n",
       "      <td>4.275609</td>\n",
       "      <td>4.241249</td>\n",
       "      <td>4.194892</td>\n",
       "      <td>4.251307</td>\n",
       "      <td>3.777344</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>0.862024859927373</td>\n",
       "      <td>1.90516376444767e-43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>4.658654</td>\n",
       "      <td>4.322857</td>\n",
       "      <td>4.697171</td>\n",
       "      <td>4.816981</td>\n",
       "      <td>4.586892</td>\n",
       "      <td>4.620245</td>\n",
       "      <td>4.640644</td>\n",
       "      <td>4.646106</td>\n",
       "      <td>4.715597</td>\n",
       "      <td>4.771581</td>\n",
       "      <td>...</td>\n",
       "      <td>4.410680</td>\n",
       "      <td>4.509940</td>\n",
       "      <td>5.482311</td>\n",
       "      <td>4.626495</td>\n",
       "      <td>4.505745</td>\n",
       "      <td>4.417440</td>\n",
       "      <td>4.359805</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>0.689133234326797</td>\n",
       "      <td>1.82706543764349e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>2.788302</td>\n",
       "      <td>3.509472</td>\n",
       "      <td>3.071359</td>\n",
       "      <td>2.453118</td>\n",
       "      <td>3.245321</td>\n",
       "      <td>3.029654</td>\n",
       "      <td>2.769487</td>\n",
       "      <td>2.762198</td>\n",
       "      <td>2.738179</td>\n",
       "      <td>2.773700</td>\n",
       "      <td>...</td>\n",
       "      <td>2.794582</td>\n",
       "      <td>3.022211</td>\n",
       "      <td>3.098520</td>\n",
       "      <td>2.858558</td>\n",
       "      <td>3.158803</td>\n",
       "      <td>2.590063</td>\n",
       "      <td>2.596908</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>0.43943150588816</td>\n",
       "      <td>4.00471354671356e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>9.294971</td>\n",
       "      <td>9.078311</td>\n",
       "      <td>9.836819</td>\n",
       "      <td>10.287786</td>\n",
       "      <td>9.612483</td>\n",
       "      <td>10.320160</td>\n",
       "      <td>10.118597</td>\n",
       "      <td>10.629440</td>\n",
       "      <td>9.767962</td>\n",
       "      <td>10.012075</td>\n",
       "      <td>...</td>\n",
       "      <td>10.088363</td>\n",
       "      <td>9.523195</td>\n",
       "      <td>10.219047</td>\n",
       "      <td>9.810799</td>\n",
       "      <td>9.909992</td>\n",
       "      <td>9.223608</td>\n",
       "      <td>9.199473</td>\n",
       "      <td>FGR</td>\n",
       "      <td>0.935942743827098</td>\n",
       "      <td>8.78762646249344e-66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288597</th>\n",
       "      <td>0.312633</td>\n",
       "      <td>0.339987</td>\n",
       "      <td>0.074365</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.047952</td>\n",
       "      <td>0.182959</td>\n",
       "      <td>0.223352</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.109893</td>\n",
       "      <td>0.112489</td>\n",
       "      <td>...</td>\n",
       "      <td>0.165031</td>\n",
       "      <td>0.300381</td>\n",
       "      <td>0.189985</td>\n",
       "      <td>0.218882</td>\n",
       "      <td>0.096484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.195488</td>\n",
       "      <td>AC234782.4</td>\n",
       "      <td>0.2517017045725</td>\n",
       "      <td>0.00242557060235678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288598</th>\n",
       "      <td>0.425238</td>\n",
       "      <td>1.851654</td>\n",
       "      <td>0.338553</td>\n",
       "      <td>0.448594</td>\n",
       "      <td>1.598257</td>\n",
       "      <td>0.225291</td>\n",
       "      <td>0.223352</td>\n",
       "      <td>0.246822</td>\n",
       "      <td>0.707687</td>\n",
       "      <td>0.216838</td>\n",
       "      <td>...</td>\n",
       "      <td>0.710390</td>\n",
       "      <td>0.269932</td>\n",
       "      <td>0.611473</td>\n",
       "      <td>0.926814</td>\n",
       "      <td>1.166692</td>\n",
       "      <td>1.295676</td>\n",
       "      <td>1.499990</td>\n",
       "      <td>AL354833.1</td>\n",
       "      <td>0.420543151899831</td>\n",
       "      <td>1.70069109830835e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288600</th>\n",
       "      <td>0.028802</td>\n",
       "      <td>0.034439</td>\n",
       "      <td>0.037662</td>\n",
       "      <td>0.047057</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.074184</td>\n",
       "      <td>0.112489</td>\n",
       "      <td>...</td>\n",
       "      <td>0.165031</td>\n",
       "      <td>0.036635</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076710</td>\n",
       "      <td>0.096484</td>\n",
       "      <td>0.050308</td>\n",
       "      <td>0.081401</td>\n",
       "      <td>AL354833.2</td>\n",
       "      <td>0.591823204966647</td>\n",
       "      <td>7.03551034285073e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288602</th>\n",
       "      <td>0.358735</td>\n",
       "      <td>0.100945</td>\n",
       "      <td>0.308056</td>\n",
       "      <td>0.301105</td>\n",
       "      <td>0.182927</td>\n",
       "      <td>0.306400</td>\n",
       "      <td>0.504476</td>\n",
       "      <td>0.612279</td>\n",
       "      <td>0.178764</td>\n",
       "      <td>0.491040</td>\n",
       "      <td>...</td>\n",
       "      <td>0.347866</td>\n",
       "      <td>0.470701</td>\n",
       "      <td>0.707862</td>\n",
       "      <td>0.317011</td>\n",
       "      <td>0.215845</td>\n",
       "      <td>0.050308</td>\n",
       "      <td>0.334798</td>\n",
       "      <td>C8orf44-SGK3</td>\n",
       "      <td>0.563582793344071</td>\n",
       "      <td>2.33295900431294e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288604</th>\n",
       "      <td>0.718316</td>\n",
       "      <td>1.334566</td>\n",
       "      <td>1.077822</td>\n",
       "      <td>2.334844</td>\n",
       "      <td>0.575356</td>\n",
       "      <td>1.598438</td>\n",
       "      <td>1.648420</td>\n",
       "      <td>1.852902</td>\n",
       "      <td>1.020715</td>\n",
       "      <td>0.950509</td>\n",
       "      <td>...</td>\n",
       "      <td>1.330485</td>\n",
       "      <td>1.375494</td>\n",
       "      <td>1.536126</td>\n",
       "      <td>1.138441</td>\n",
       "      <td>0.816240</td>\n",
       "      <td>0.817988</td>\n",
       "      <td>0.925766</td>\n",
       "      <td>AC015871.8</td>\n",
       "      <td>0.808407789687762</td>\n",
       "      <td>2.91317497737892e-34</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23063 rows × 146 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 S00016-Ja001T-TRGa  S00020-Ja003T-TRGa  S00024-Ja003T-TRGa   \n",
       "ENSG00000000003            0.358735            0.420317            0.212498  \\\n",
       "ENSG00000000419            3.933231            4.017549            4.351443   \n",
       "ENSG00000000457            4.658654            4.322857            4.697171   \n",
       "ENSG00000000460            2.788302            3.509472            3.071359   \n",
       "ENSG00000000938            9.294971            9.078311            9.836819   \n",
       "...                             ...                 ...                 ...   \n",
       "ENSG00000288597            0.312633            0.339987            0.074365   \n",
       "ENSG00000288598            0.425238            1.851654            0.338553   \n",
       "ENSG00000288600            0.028802            0.034439            0.037662   \n",
       "ENSG00000288602            0.358735            0.100945            0.308056   \n",
       "ENSG00000288604            0.718316            1.334566            1.077822   \n",
       "\n",
       "                 S00027-Ja003T-TRGa  S00028-Ja001T-TRGa  S00030-Ja003T-TRGa   \n",
       "ENSG00000000003            0.000000            0.420037            0.000000  \\\n",
       "ENSG00000000419            4.102646            4.466349            3.650819   \n",
       "ENSG00000000457            4.816981            4.586892            4.620245   \n",
       "ENSG00000000460            2.453118            3.245321            3.029654   \n",
       "ENSG00000000938           10.287786            9.612483           10.320160   \n",
       "...                             ...                 ...                 ...   \n",
       "ENSG00000288597            0.000000            0.047952            0.182959   \n",
       "ENSG00000288598            0.448594            1.598257            0.225291   \n",
       "ENSG00000288600            0.047057            0.000000            0.000000   \n",
       "ENSG00000288602            0.301105            0.182927            0.306400   \n",
       "ENSG00000288604            2.334844            0.575356            1.598438   \n",
       "\n",
       "                 S00033-Ja001T-TRGa  S00033-Ja003T-TRGa  S00034-Ja005T-TRGa   \n",
       "ENSG00000000003            0.000000            0.044182            0.037569  \\\n",
       "ENSG00000000419            4.111353            4.330951            4.395805   \n",
       "ENSG00000000457            4.640644            4.646106            4.715597   \n",
       "ENSG00000000460            2.769487            2.762198            2.738179   \n",
       "ENSG00000000938           10.118597           10.629440            9.767962   \n",
       "...                             ...                 ...                 ...   \n",
       "ENSG00000288597            0.223352            0.000000            0.109893   \n",
       "ENSG00000288598            0.223352            0.246822            0.707687   \n",
       "ENSG00000288600            0.000000            0.000000            0.074184   \n",
       "ENSG00000288602            0.504476            0.612279            0.178764   \n",
       "ENSG00000288604            1.648420            1.852902            1.020715   \n",
       "\n",
       "                 S00033-Ja005T-TRGa  ...  S00094-Ja005T-TRGa   \n",
       "ENSG00000000003            0.266312  ...            0.240967  \\\n",
       "ENSG00000000419            4.293557  ...            3.874222   \n",
       "ENSG00000000457            4.771581  ...            4.410680   \n",
       "ENSG00000000460            2.773700  ...            2.794582   \n",
       "ENSG00000000938           10.012075  ...           10.088363   \n",
       "...                             ...  ...                 ...   \n",
       "ENSG00000288597            0.112489  ...            0.165031   \n",
       "ENSG00000288598            0.216838  ...            0.710390   \n",
       "ENSG00000288600            0.112489  ...            0.165031   \n",
       "ENSG00000288602            0.491040  ...            0.347866   \n",
       "ENSG00000288604            0.950509  ...            1.330485   \n",
       "\n",
       "                 S00095-Ja005T-TRGa  S00096-Ja005T-TRGa  S00097-Ja003T-TRGa   \n",
       "ENSG00000000003            0.269932            0.189985            0.285038  \\\n",
       "ENSG00000000419            3.827185            4.275609            4.241249   \n",
       "ENSG00000000457            4.509940            5.482311            4.626495   \n",
       "ENSG00000000460            3.022211            3.098520            2.858558   \n",
       "ENSG00000000938            9.523195           10.219047            9.810799   \n",
       "...                             ...                 ...                 ...   \n",
       "ENSG00000288597            0.300381            0.189985            0.218882   \n",
       "ENSG00000288598            0.269932            0.611473            0.926814   \n",
       "ENSG00000288600            0.036635            0.000000            0.076710   \n",
       "ENSG00000288602            0.470701            0.707862            0.317011   \n",
       "ENSG00000288604            1.375494            1.536126            1.138441   \n",
       "\n",
       "                 S00099-Ja005T-TRGa  S00104-Ja003T-TRGa  S00106-Ja003T-TRGa   \n",
       "ENSG00000000003            0.352373            0.581883            0.000000  \\\n",
       "ENSG00000000419            4.194892            4.251307            3.777344   \n",
       "ENSG00000000457            4.505745            4.417440            4.359805   \n",
       "ENSG00000000460            3.158803            2.590063            2.596908   \n",
       "ENSG00000000938            9.909992            9.223608            9.199473   \n",
       "...                             ...                 ...                 ...   \n",
       "ENSG00000288597            0.096484            0.000000            0.195488   \n",
       "ENSG00000288598            1.166692            1.295676            1.499990   \n",
       "ENSG00000288600            0.096484            0.050308            0.081401   \n",
       "ENSG00000288602            0.215845            0.050308            0.334798   \n",
       "ENSG00000288604            0.816240            0.817988            0.925766   \n",
       "\n",
       "                    gene_name         membership               p.value  \n",
       "ENSG00000000003        TSPAN6  0.634714068658144  1.70946329554292e-17  \n",
       "ENSG00000000419          DPM1  0.862024859927373  1.90516376444767e-43  \n",
       "ENSG00000000457         SCYL3  0.689133234326797  1.82706543764349e-21  \n",
       "ENSG00000000460      C1orf112   0.43943150588816  4.00471354671356e-08  \n",
       "ENSG00000000938           FGR  0.935942743827098  8.78762646249344e-66  \n",
       "...                       ...                ...                   ...  \n",
       "ENSG00000288597    AC234782.4    0.2517017045725   0.00242557060235678  \n",
       "ENSG00000288598    AL354833.1  0.420543151899831  1.70069109830835e-07  \n",
       "ENSG00000288600    AL354833.2  0.591823204966647  7.03551034285073e-15  \n",
       "ENSG00000288602  C8orf44-SGK3  0.563582793344071  2.33295900431294e-13  \n",
       "ENSG00000288604    AC015871.8  0.808407789687762  2.91317497737892e-34  \n",
       "\n",
       "[23063 rows x 146 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bulk_RNA = read(\"Logcpm_143_23063\", \"txt\")\n",
    "print(\"bulk RNA: \", bulk_RNA, \"\\n\\n\") # Note that here featres are rows \n",
    "\n",
    "bulk_RNA2 = read(\"module.gene.membership\", \"tsv\") \n",
    "print(\"bulk RNA: \", bulk_RNA2, \"\\n\\n\") # Maps gene_ids to gene_names\n",
    "\n",
    "bulkRNA_joined = bulk_RNA.to_df().join(bulk_RNA2.to_df().set_index(\"gene_id\")) # Join the two data frames based on \"gene_id\"\n",
    "bulkRNA_joined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "Cannot index by location index with a non-integer key",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m bulkRNA_joined\u001b[39m.\u001b[39;49miloc[\u001b[39m\"\u001b[39;49m\u001b[39mgene_name\u001b[39;49m\u001b[39m\"\u001b[39;49m \u001b[39m==\u001b[39;49m \u001b[39m\"\u001b[39;49m\u001b[39mTSPAN6\u001b[39;49m\u001b[39m\"\u001b[39;49m]\n",
      "File \u001b[0;32m/opt/anaconda3/envs/thesis/lib/python3.10/site-packages/pandas/core/indexing.py:1103\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1100\u001b[0m axis \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39maxis \u001b[39mor\u001b[39;00m \u001b[39m0\u001b[39m\n\u001b[1;32m   1102\u001b[0m maybe_callable \u001b[39m=\u001b[39m com\u001b[39m.\u001b[39mapply_if_callable(key, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mobj)\n\u001b[0;32m-> 1103\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_getitem_axis(maybe_callable, axis\u001b[39m=\u001b[39;49maxis)\n",
      "File \u001b[0;32m/opt/anaconda3/envs/thesis/lib/python3.10/site-packages/pandas/core/indexing.py:1653\u001b[0m, in \u001b[0;36m_iLocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1651\u001b[0m key \u001b[39m=\u001b[39m item_from_zerodim(key)\n\u001b[1;32m   1652\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m is_integer(key):\n\u001b[0;32m-> 1653\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mCannot index by location index with a non-integer key\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m   1655\u001b[0m \u001b[39m# validate the location\u001b[39;00m\n\u001b[1;32m   1656\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_validate_integer(key, axis)\n",
      "\u001b[0;31mTypeError\u001b[0m: Cannot index by location index with a non-integer key"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 0.0 , 15.981462 ]\n"
     ]
    }
   ],
   "source": [
    "summary_bulkRNA = {\"feature\" : bulkRNA_joined[\"gene_name\"],\n",
    "           \"min\" : bulk_RNA.X.min(axis = 1),\n",
    "           \"max\" : bulk_RNA.X.max(axis = 1),\n",
    "           \"mean\" : bulk_RNA.X.mean(axis = 1),\n",
    "           \"var\" : bulk_RNA.X.var(axis = 1)}\n",
    "\n",
    "df_bulkRNA= pd.DataFrame(summary_bulkRNA)\n",
    "\n",
    "print(\"[\", bulk_RNA.X.min(),\",\", bulk_RNA.X.max(), \"]\") # Overall range of proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S00016-Ja001T-TRGa</th>\n",
       "      <th>S00020-Ja003T-TRGa</th>\n",
       "      <th>S00024-Ja003T-TRGa</th>\n",
       "      <th>S00027-Ja003T-TRGa</th>\n",
       "      <th>S00028-Ja001T-TRGa</th>\n",
       "      <th>S00030-Ja003T-TRGa</th>\n",
       "      <th>S00033-Ja001T-TRGa</th>\n",
       "      <th>S00033-Ja003T-TRGa</th>\n",
       "      <th>S00034-Ja005T-TRGa</th>\n",
       "      <th>S00033-Ja005T-TRGa</th>\n",
       "      <th>...</th>\n",
       "      <th>S00081-Ja001T-TRGa</th>\n",
       "      <th>S00081-Ja005T-TRGa</th>\n",
       "      <th>S00082-Ja001T-TRGa</th>\n",
       "      <th>S00094-Ja005T-TRGa</th>\n",
       "      <th>S00095-Ja005T-TRGa</th>\n",
       "      <th>S00096-Ja005T-TRGa</th>\n",
       "      <th>S00097-Ja003T-TRGa</th>\n",
       "      <th>S00099-Ja005T-TRGa</th>\n",
       "      <th>S00104-Ja003T-TRGa</th>\n",
       "      <th>S00106-Ja003T-TRGa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>0.358735</td>\n",
       "      <td>0.420317</td>\n",
       "      <td>0.212498</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.420037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044182</td>\n",
       "      <td>0.037569</td>\n",
       "      <td>0.266312</td>\n",
       "      <td>...</td>\n",
       "      <td>0.236062</td>\n",
       "      <td>0.181513</td>\n",
       "      <td>0.382617</td>\n",
       "      <td>0.240967</td>\n",
       "      <td>0.269932</td>\n",
       "      <td>0.189985</td>\n",
       "      <td>0.285038</td>\n",
       "      <td>0.352373</td>\n",
       "      <td>0.581883</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>3.933231</td>\n",
       "      <td>4.017549</td>\n",
       "      <td>4.351443</td>\n",
       "      <td>4.102646</td>\n",
       "      <td>4.466349</td>\n",
       "      <td>3.650819</td>\n",
       "      <td>4.111353</td>\n",
       "      <td>4.330951</td>\n",
       "      <td>4.395805</td>\n",
       "      <td>4.293557</td>\n",
       "      <td>...</td>\n",
       "      <td>4.074801</td>\n",
       "      <td>4.310133</td>\n",
       "      <td>4.190965</td>\n",
       "      <td>3.874222</td>\n",
       "      <td>3.827185</td>\n",
       "      <td>4.275609</td>\n",
       "      <td>4.241249</td>\n",
       "      <td>4.194892</td>\n",
       "      <td>4.251307</td>\n",
       "      <td>3.777344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>4.658654</td>\n",
       "      <td>4.322857</td>\n",
       "      <td>4.697171</td>\n",
       "      <td>4.816981</td>\n",
       "      <td>4.586892</td>\n",
       "      <td>4.620245</td>\n",
       "      <td>4.640644</td>\n",
       "      <td>4.646106</td>\n",
       "      <td>4.715597</td>\n",
       "      <td>4.771581</td>\n",
       "      <td>...</td>\n",
       "      <td>4.734925</td>\n",
       "      <td>4.968107</td>\n",
       "      <td>4.671478</td>\n",
       "      <td>4.410680</td>\n",
       "      <td>4.509940</td>\n",
       "      <td>5.482311</td>\n",
       "      <td>4.626495</td>\n",
       "      <td>4.505745</td>\n",
       "      <td>4.417440</td>\n",
       "      <td>4.359805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>2.788302</td>\n",
       "      <td>3.509472</td>\n",
       "      <td>3.071359</td>\n",
       "      <td>2.453118</td>\n",
       "      <td>3.245321</td>\n",
       "      <td>3.029654</td>\n",
       "      <td>2.769487</td>\n",
       "      <td>2.762198</td>\n",
       "      <td>2.738179</td>\n",
       "      <td>2.773700</td>\n",
       "      <td>...</td>\n",
       "      <td>3.121597</td>\n",
       "      <td>2.885727</td>\n",
       "      <td>3.201638</td>\n",
       "      <td>2.794582</td>\n",
       "      <td>3.022211</td>\n",
       "      <td>3.098520</td>\n",
       "      <td>2.858558</td>\n",
       "      <td>3.158803</td>\n",
       "      <td>2.590063</td>\n",
       "      <td>2.596908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>9.294971</td>\n",
       "      <td>9.078311</td>\n",
       "      <td>9.836819</td>\n",
       "      <td>10.287786</td>\n",
       "      <td>9.612483</td>\n",
       "      <td>10.320160</td>\n",
       "      <td>10.118597</td>\n",
       "      <td>10.629440</td>\n",
       "      <td>9.767962</td>\n",
       "      <td>10.012075</td>\n",
       "      <td>...</td>\n",
       "      <td>9.839436</td>\n",
       "      <td>9.990548</td>\n",
       "      <td>9.713717</td>\n",
       "      <td>10.088363</td>\n",
       "      <td>9.523195</td>\n",
       "      <td>10.219047</td>\n",
       "      <td>9.810799</td>\n",
       "      <td>9.909992</td>\n",
       "      <td>9.223608</td>\n",
       "      <td>9.199473</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 143 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 S00016-Ja001T-TRGa  S00020-Ja003T-TRGa  S00024-Ja003T-TRGa   \n",
       "ENSG00000000003            0.358735            0.420317            0.212498  \\\n",
       "ENSG00000000419            3.933231            4.017549            4.351443   \n",
       "ENSG00000000457            4.658654            4.322857            4.697171   \n",
       "ENSG00000000460            2.788302            3.509472            3.071359   \n",
       "ENSG00000000938            9.294971            9.078311            9.836819   \n",
       "\n",
       "                 S00027-Ja003T-TRGa  S00028-Ja001T-TRGa  S00030-Ja003T-TRGa   \n",
       "ENSG00000000003            0.000000            0.420037            0.000000  \\\n",
       "ENSG00000000419            4.102646            4.466349            3.650819   \n",
       "ENSG00000000457            4.816981            4.586892            4.620245   \n",
       "ENSG00000000460            2.453118            3.245321            3.029654   \n",
       "ENSG00000000938           10.287786            9.612483           10.320160   \n",
       "\n",
       "                 S00033-Ja001T-TRGa  S00033-Ja003T-TRGa  S00034-Ja005T-TRGa   \n",
       "ENSG00000000003            0.000000            0.044182            0.037569  \\\n",
       "ENSG00000000419            4.111353            4.330951            4.395805   \n",
       "ENSG00000000457            4.640644            4.646106            4.715597   \n",
       "ENSG00000000460            2.769487            2.762198            2.738179   \n",
       "ENSG00000000938           10.118597           10.629440            9.767962   \n",
       "\n",
       "                 S00033-Ja005T-TRGa  ...  S00081-Ja001T-TRGa   \n",
       "ENSG00000000003            0.266312  ...            0.236062  \\\n",
       "ENSG00000000419            4.293557  ...            4.074801   \n",
       "ENSG00000000457            4.771581  ...            4.734925   \n",
       "ENSG00000000460            2.773700  ...            3.121597   \n",
       "ENSG00000000938           10.012075  ...            9.839436   \n",
       "\n",
       "                 S00081-Ja005T-TRGa  S00082-Ja001T-TRGa  S00094-Ja005T-TRGa   \n",
       "ENSG00000000003            0.181513            0.382617            0.240967  \\\n",
       "ENSG00000000419            4.310133            4.190965            3.874222   \n",
       "ENSG00000000457            4.968107            4.671478            4.410680   \n",
       "ENSG00000000460            2.885727            3.201638            2.794582   \n",
       "ENSG00000000938            9.990548            9.713717           10.088363   \n",
       "\n",
       "                 S00095-Ja005T-TRGa  S00096-Ja005T-TRGa  S00097-Ja003T-TRGa   \n",
       "ENSG00000000003            0.269932            0.189985            0.285038  \\\n",
       "ENSG00000000419            3.827185            4.275609            4.241249   \n",
       "ENSG00000000457            4.509940            5.482311            4.626495   \n",
       "ENSG00000000460            3.022211            3.098520            2.858558   \n",
       "ENSG00000000938            9.523195           10.219047            9.810799   \n",
       "\n",
       "                 S00099-Ja005T-TRGa  S00104-Ja003T-TRGa  S00106-Ja003T-TRGa  \n",
       "ENSG00000000003            0.352373            0.581883            0.000000  \n",
       "ENSG00000000419            4.194892            4.251307            3.777344  \n",
       "ENSG00000000457            4.505745            4.417440            4.359805  \n",
       "ENSG00000000460            3.158803            2.590063            2.596908  \n",
       "ENSG00000000938            9.909992            9.223608            9.199473  \n",
       "\n",
       "[5 rows x 143 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bulk_RNA.to_df().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open ('bulkRNA_feature_names.txt', 'w') as file: # Save protein names\n",
    "    for name in bulkRNA_joined[\"gene_name\"]:\n",
    "        file.write(name + \"\\n\")  \n",
    "\n",
    "df_bulkRNA.to_excel(\"summary_bulkRNA.xlsx\") # Save summary\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Luminex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Luminex:  AnnData object with n_obs × n_vars = 349 × 56 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "luminex = read(\"Oxford data_output_combined\", \"xlsx\", \"All data\")\n",
    "print(\"Luminex: \", luminex, \"\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 0.0 , 12019000.0 ]\n"
     ]
    }
   ],
   "source": [
    "summary_luminex = {\"feature\" : luminex.var_names[5:],\n",
    "           \"min\" : luminex.X[:,5:].astype(float).min(axis = 0),\n",
    "           \"max\" : luminex.X[:,5:].astype(float).max(axis = 0),\n",
    "           \"mean\" : luminex.X[:,5:].astype(float).mean(axis = 0),\n",
    "           \"var\" : luminex.X[:,5:].astype(float).var(axis = 0)}\n",
    "\n",
    "df_luminex= pd.DataFrame(summary_luminex)\n",
    "\n",
    "# Overall range of proteins\n",
    "print(\"[\", luminex.X[:,5:].astype(float).min(),\",\",luminex.X[:,5:].astype(float).max(), \"]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>severity</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>BMI</th>\n",
       "      <th>dexamethasone</th>\n",
       "      <th>CCL18/PARC (BR33) (33) low</th>\n",
       "      <th>Lactoferrin (BR36) (36) high</th>\n",
       "      <th>Lipocalin-2/NGAL (BR21) (21) high</th>\n",
       "      <th>Myeloperoxidase/MPO (BR53) (53) high</th>\n",
       "      <th>CCL2/JE/MCP-1 (BR25) (25) high</th>\n",
       "      <th>...</th>\n",
       "      <th>IFN-alpha (BR63) (63) high</th>\n",
       "      <th>IL-2 (BR43) (43) high</th>\n",
       "      <th>IL-5 (BR53) (53) high</th>\n",
       "      <th>IL-8/CXCL8 (BR48) (48) high</th>\n",
       "      <th>IL-12 p70 (BR56) (56) high</th>\n",
       "      <th>IL-15 (BR52) (52) high</th>\n",
       "      <th>IL-23 (BR76) (76) high</th>\n",
       "      <th>IL-33 (BR14) (14) high</th>\n",
       "      <th>Oncostatin M/OSM (BR30) (30) high</th>\n",
       "      <th>TREM-1 (BR65) (65) high</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S00029-Ja005E-PMCdb</th>\n",
       "      <td>COVID-critical</td>\n",
       "      <td>F</td>\n",
       "      <td>46</td>\n",
       "      <td>21.4</td>\n",
       "      <td>False</td>\n",
       "      <td>42080.95</td>\n",
       "      <td>15359.95</td>\n",
       "      <td>528930.72</td>\n",
       "      <td>101595.14</td>\n",
       "      <td>185.96</td>\n",
       "      <td>...</td>\n",
       "      <td>2.74</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.71</td>\n",
       "      <td>11.53</td>\n",
       "      <td>15.05</td>\n",
       "      <td>10.5</td>\n",
       "      <td>151.94</td>\n",
       "      <td>1.44</td>\n",
       "      <td>0.0</td>\n",
       "      <td>327.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00029-Ja001E-PMCdb</th>\n",
       "      <td>COVID-critical</td>\n",
       "      <td>F</td>\n",
       "      <td>46</td>\n",
       "      <td>21.4</td>\n",
       "      <td>False</td>\n",
       "      <td>47991.24</td>\n",
       "      <td>96194.78</td>\n",
       "      <td>467186.02</td>\n",
       "      <td>39638.1</td>\n",
       "      <td>1966.97</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.35</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>728.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00052-Ja005E-PMCdb</th>\n",
       "      <td>COVID-critical</td>\n",
       "      <td>F</td>\n",
       "      <td>41</td>\n",
       "      <td>30.0</td>\n",
       "      <td>False</td>\n",
       "      <td>76407.98</td>\n",
       "      <td>45607.18</td>\n",
       "      <td>51840.97</td>\n",
       "      <td>88627.44</td>\n",
       "      <td>174.56</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.74</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.38</td>\n",
       "      <td>0.0</td>\n",
       "      <td>269.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00109-Ja005E-PMCdb</th>\n",
       "      <td>COVID-critical</td>\n",
       "      <td>F</td>\n",
       "      <td>52</td>\n",
       "      <td>nan</td>\n",
       "      <td>False</td>\n",
       "      <td>68297.35</td>\n",
       "      <td>64517.42</td>\n",
       "      <td>27631.27</td>\n",
       "      <td>77299.95</td>\n",
       "      <td>2148.9</td>\n",
       "      <td>...</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>63.57</td>\n",
       "      <td>96.28</td>\n",
       "      <td>19.37</td>\n",
       "      <td>524.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>246.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00099-Ja005E-PMCdb</th>\n",
       "      <td>COVID-critical</td>\n",
       "      <td>F</td>\n",
       "      <td>52</td>\n",
       "      <td>35.0</td>\n",
       "      <td>False</td>\n",
       "      <td>171822.5</td>\n",
       "      <td>181611.29</td>\n",
       "      <td>245606.35</td>\n",
       "      <td>166649.32</td>\n",
       "      <td>1009.95</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.96</td>\n",
       "      <td>8.35</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>955.97</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           severity sex age   BMI dexamethasone   \n",
       "S00029-Ja005E-PMCdb  COVID-critical   F  46  21.4         False  \\\n",
       "S00029-Ja001E-PMCdb  COVID-critical   F  46  21.4         False   \n",
       "S00052-Ja005E-PMCdb  COVID-critical   F  41  30.0         False   \n",
       "S00109-Ja005E-PMCdb  COVID-critical   F  52   nan         False   \n",
       "S00099-Ja005E-PMCdb  COVID-critical   F  52  35.0         False   \n",
       "\n",
       "                    CCL18/PARC (BR33) (33) low Lactoferrin (BR36) (36) high   \n",
       "S00029-Ja005E-PMCdb                   42080.95                     15359.95  \\\n",
       "S00029-Ja001E-PMCdb                   47991.24                     96194.78   \n",
       "S00052-Ja005E-PMCdb                   76407.98                     45607.18   \n",
       "S00109-Ja005E-PMCdb                   68297.35                     64517.42   \n",
       "S00099-Ja005E-PMCdb                   171822.5                    181611.29   \n",
       "\n",
       "                    Lipocalin-2/NGAL (BR21) (21) high   \n",
       "S00029-Ja005E-PMCdb                         528930.72  \\\n",
       "S00029-Ja001E-PMCdb                         467186.02   \n",
       "S00052-Ja005E-PMCdb                          51840.97   \n",
       "S00109-Ja005E-PMCdb                          27631.27   \n",
       "S00099-Ja005E-PMCdb                         245606.35   \n",
       "\n",
       "                    Myeloperoxidase/MPO (BR53) (53) high   \n",
       "S00029-Ja005E-PMCdb                            101595.14  \\\n",
       "S00029-Ja001E-PMCdb                              39638.1   \n",
       "S00052-Ja005E-PMCdb                             88627.44   \n",
       "S00109-Ja005E-PMCdb                             77299.95   \n",
       "S00099-Ja005E-PMCdb                            166649.32   \n",
       "\n",
       "                    CCL2/JE/MCP-1 (BR25) (25) high  ...   \n",
       "S00029-Ja005E-PMCdb                         185.96  ...  \\\n",
       "S00029-Ja001E-PMCdb                        1966.97  ...   \n",
       "S00052-Ja005E-PMCdb                         174.56  ...   \n",
       "S00109-Ja005E-PMCdb                         2148.9  ...   \n",
       "S00099-Ja005E-PMCdb                        1009.95  ...   \n",
       "\n",
       "                    IFN-alpha (BR63) (63) high IL-2 (BR43) (43) high   \n",
       "S00029-Ja005E-PMCdb                       2.74                   0.0  \\\n",
       "S00029-Ja001E-PMCdb                        0.0                  9.42   \n",
       "S00052-Ja005E-PMCdb                        0.0                   0.0   \n",
       "S00109-Ja005E-PMCdb                       1.08                   0.0   \n",
       "S00099-Ja005E-PMCdb                        0.0                   0.0   \n",
       "\n",
       "                    IL-5 (BR53) (53) high IL-8/CXCL8 (BR48) (48) high   \n",
       "S00029-Ja005E-PMCdb                  5.71                       11.53  \\\n",
       "S00029-Ja001E-PMCdb                   0.0                       14.35   \n",
       "S00052-Ja005E-PMCdb                   0.0                        3.74   \n",
       "S00109-Ja005E-PMCdb                   0.0                       63.57   \n",
       "S00099-Ja005E-PMCdb                  5.96                        8.35   \n",
       "\n",
       "                    IL-12 p70 (BR56) (56) high IL-15 (BR52) (52) high   \n",
       "S00029-Ja005E-PMCdb                      15.05                   10.5  \\\n",
       "S00029-Ja001E-PMCdb                        0.0                  19.25   \n",
       "S00052-Ja005E-PMCdb                        0.0                   0.38   \n",
       "S00109-Ja005E-PMCdb                      96.28                  19.37   \n",
       "S00099-Ja005E-PMCdb                        0.0                   8.98   \n",
       "\n",
       "                    IL-23 (BR76) (76) high IL-33 (BR14) (14) high   \n",
       "S00029-Ja005E-PMCdb                 151.94                   1.44  \\\n",
       "S00029-Ja001E-PMCdb                    0.0                    0.0   \n",
       "S00052-Ja005E-PMCdb                    0.0                   2.38   \n",
       "S00109-Ja005E-PMCdb                 524.06                    0.0   \n",
       "S00099-Ja005E-PMCdb                    0.0                    0.0   \n",
       "\n",
       "                    Oncostatin M/OSM (BR30) (30) high TREM-1 (BR65) (65) high  \n",
       "S00029-Ja005E-PMCdb                               0.0                  327.96  \n",
       "S00029-Ja001E-PMCdb                               0.0                  728.84  \n",
       "S00052-Ja005E-PMCdb                               0.0                   269.0  \n",
       "S00109-Ja005E-PMCdb                               0.0                  246.34  \n",
       "S00099-Ja005E-PMCdb                               0.0                  955.97  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "luminex.to_df().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open ('luminex_feature_names.txt', 'w') as file:  \n",
    "    for name in luminex.var_names:\n",
    "        file.write(name + \"\\n\") \n",
    "\n",
    "df_luminex.to_excel(\"summary_Luminex.xlsx\") # Save summary"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cyTOF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/thesis/lib/python3.10/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cyTOF:  AnnData object with n_obs × n_vars = 7118158 × 48\n",
      "    obs: 'sample_id', 'condition', 'patient_id', 'batch', 'cellID', 'COMBAT_ID_Time', 'CyTOF_priority', 'major_cell_type', 'fine_cluster_id'\n",
      "    var: 'channel_name', 'marker_name', 'marker_class'\n",
      "    uns: 'SOM_codes', 'X_name', 'cluster_codes', 'cofactor', 'experiment_info'\n",
      "    obsm: 'TSNE', 'UMAP' \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cytof = read(\"cytof_full\", \"h5ad\")\n",
    "print(\"cyTOF: \", cytof, \"\\n\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ -6.436475 , 24.303553 ]\n"
     ]
    }
   ],
   "source": [
    "summary_cytof = {\"feature\" : cytof.var_names,\n",
    "           \"min\" : cytof.X.min(axis = 0),\n",
    "           \"max\" : cytof.X.max(axis = 0),\n",
    "           \"mean\" : cytof.X.mean(axis = 0),\n",
    "           \"var\" : cytof.X.var(axis = 0)\n",
    "           }\n",
    "\n",
    "df_cytof= pd.DataFrame(summary_cytof)\n",
    "\n",
    "# Overall range of the proteins\n",
    "print(\"[\", cytof.X.min(),\",\",cytof.X.max(), \"]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD16</th>\n",
       "      <th>CD19</th>\n",
       "      <th>CD3</th>\n",
       "      <th>IgG</th>\n",
       "      <th>CD4</th>\n",
       "      <th>HLA_DR</th>\n",
       "      <th>CTLA4</th>\n",
       "      <th>Siglec_8</th>\n",
       "      <th>CD28</th>\n",
       "      <th>Ki_67</th>\n",
       "      <th>...</th>\n",
       "      <th>KLGR1</th>\n",
       "      <th>FOXP3</th>\n",
       "      <th>CD38</th>\n",
       "      <th>CD45</th>\n",
       "      <th>CD123</th>\n",
       "      <th>CD25</th>\n",
       "      <th>CD141</th>\n",
       "      <th>CLA</th>\n",
       "      <th>CX3CR1</th>\n",
       "      <th>Event_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.002863</td>\n",
       "      <td>0.007120</td>\n",
       "      <td>3.169593</td>\n",
       "      <td>0.778884</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>0.877128</td>\n",
       "      <td>0.482226</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.489538</td>\n",
       "      <td>1.481506</td>\n",
       "      <td>...</td>\n",
       "      <td>2.412200</td>\n",
       "      <td>0.006150</td>\n",
       "      <td>0.005076</td>\n",
       "      <td>5.388708</td>\n",
       "      <td>0.032491</td>\n",
       "      <td>0.922501</td>\n",
       "      <td>0.963545</td>\n",
       "      <td>2.933704</td>\n",
       "      <td>0.333416</td>\n",
       "      <td>2.327090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.001721</td>\n",
       "      <td>2.329424</td>\n",
       "      <td>0.001770</td>\n",
       "      <td>1.906131</td>\n",
       "      <td>0.001495</td>\n",
       "      <td>4.516001</td>\n",
       "      <td>0.492553</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.296466</td>\n",
       "      <td>0.024739</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.024434</td>\n",
       "      <td>2.667199</td>\n",
       "      <td>5.177439</td>\n",
       "      <td>0.022996</td>\n",
       "      <td>0.000264</td>\n",
       "      <td>0.001856</td>\n",
       "      <td>0.826805</td>\n",
       "      <td>0.484731</td>\n",
       "      <td>2.363303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.001721</td>\n",
       "      <td>0.007947</td>\n",
       "      <td>0.231640</td>\n",
       "      <td>0.302123</td>\n",
       "      <td>2.170780</td>\n",
       "      <td>3.362985</td>\n",
       "      <td>0.226497</td>\n",
       "      <td>0.437236</td>\n",
       "      <td>0.078629</td>\n",
       "      <td>0.686172</td>\n",
       "      <td>...</td>\n",
       "      <td>0.265121</td>\n",
       "      <td>0.301607</td>\n",
       "      <td>2.895148</td>\n",
       "      <td>4.429830</td>\n",
       "      <td>0.031306</td>\n",
       "      <td>0.727448</td>\n",
       "      <td>0.631563</td>\n",
       "      <td>3.159231</td>\n",
       "      <td>0.020285</td>\n",
       "      <td>2.191561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.001721</td>\n",
       "      <td>0.130171</td>\n",
       "      <td>0.001770</td>\n",
       "      <td>1.143770</td>\n",
       "      <td>0.001495</td>\n",
       "      <td>4.477472</td>\n",
       "      <td>1.895761</td>\n",
       "      <td>0.010462</td>\n",
       "      <td>0.078629</td>\n",
       "      <td>0.024739</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016187</td>\n",
       "      <td>0.003332</td>\n",
       "      <td>3.320227</td>\n",
       "      <td>4.470901</td>\n",
       "      <td>0.013590</td>\n",
       "      <td>0.233912</td>\n",
       "      <td>0.001856</td>\n",
       "      <td>3.245132</td>\n",
       "      <td>0.020285</td>\n",
       "      <td>2.142490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.010130</td>\n",
       "      <td>0.007947</td>\n",
       "      <td>0.001770</td>\n",
       "      <td>1.599139</td>\n",
       "      <td>1.396846</td>\n",
       "      <td>3.425147</td>\n",
       "      <td>0.527159</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.472719</td>\n",
       "      <td>0.024739</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012915</td>\n",
       "      <td>1.370077</td>\n",
       "      <td>3.129075</td>\n",
       "      <td>4.142280</td>\n",
       "      <td>0.000452</td>\n",
       "      <td>1.384267</td>\n",
       "      <td>0.756194</td>\n",
       "      <td>3.142244</td>\n",
       "      <td>0.667578</td>\n",
       "      <td>2.405398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CD16      CD19       CD3       IgG       CD4    HLA_DR     CTLA4   \n",
       "0  0.002863  0.007120  3.169593  0.778884  0.000112  0.877128  0.482226  \\\n",
       "1  0.001721  2.329424  0.001770  1.906131  0.001495  4.516001  0.492553   \n",
       "2  0.001721  0.007947  0.231640  0.302123  2.170780  3.362985  0.226497   \n",
       "3  0.001721  0.130171  0.001770  1.143770  0.001495  4.477472  1.895761   \n",
       "4  1.010130  0.007947  0.001770  1.599139  1.396846  3.425147  0.527159   \n",
       "\n",
       "   Siglec_8      CD28     Ki_67  ...     KLGR1     FOXP3      CD38      CD45   \n",
       "0  0.000000  0.489538  1.481506  ...  2.412200  0.006150  0.005076  5.388708  \\\n",
       "1  0.000000  0.296466  0.024739  ...  0.000000  1.024434  2.667199  5.177439   \n",
       "2  0.437236  0.078629  0.686172  ...  0.265121  0.301607  2.895148  4.429830   \n",
       "3  0.010462  0.078629  0.024739  ...  0.016187  0.003332  3.320227  4.470901   \n",
       "4  0.000000  0.472719  0.024739  ...  0.012915  1.370077  3.129075  4.142280   \n",
       "\n",
       "      CD123      CD25     CD141       CLA    CX3CR1  Event_length  \n",
       "0  0.032491  0.922501  0.963545  2.933704  0.333416      2.327090  \n",
       "1  0.022996  0.000264  0.001856  0.826805  0.484731      2.363303  \n",
       "2  0.031306  0.727448  0.631563  3.159231  0.020285      2.191561  \n",
       "3  0.013590  0.233912  0.001856  3.245132  0.020285      2.142490  \n",
       "4  0.000452  1.384267  0.756194  3.142244  0.667578      2.405398  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cytof.to_df().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ -6.436475 , 24.303553 ]\n"
     ]
    }
   ],
   "source": [
    "with open ('cyTOF_feature_names.txt', 'w') as file:  \n",
    "    for name in cytof.var_names:\n",
    "        file.write(name + \"\\n\") \n",
    "\n",
    "df_cytof.to_excel(\"summary_cytof.xlsx\") # Save summary"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FACS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FACS:  AnnData object with n_obs × n_vars = 131920 × 12\n",
      "    obs: 'fcs_file', 'sample_id', 'condition', 'patient_id', 'cluster_id'\n",
      "    var: 'channel_name', 'marker_name', 'marker_class', 'used_for_clustering'\n",
      "    uns: 'SOM_codes', 'X_name', 'cluster_codes', 'experiment_info'\n",
      "    obsm: 'TSNE', 'UMAP'\n",
      "    layers: 'exprs' \n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/thesis/lib/python3.10/site-packages/anndata/_core/anndata.py:1828: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "facs = read(\"facs_full\", \"h5ad\")\n",
    "print(\"FACS: \", facs, \"\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ -6608.2285 , 178222.64 ]\n"
     ]
    }
   ],
   "source": [
    "summary_facs = {\"feature\" : facs.var_names,\n",
    "           \"min\" : facs.X.min(axis = 0),\n",
    "           \"max\" : facs.X.max(axis = 0),\n",
    "           \"mean\" : facs.X.mean(axis = 0),\n",
    "           \"var\" : facs.X.var(axis = 0)}\n",
    "\n",
    "df_facs= pd.DataFrame(summary_facs)\n",
    "\n",
    "# Overall range of proteins\n",
    "print(\"[\", facs.X.min(),\",\",facs.X.max(), \"]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CXCR3</th>\n",
       "      <th>CCR4</th>\n",
       "      <th>CD45RA</th>\n",
       "      <th>HLA-DR</th>\n",
       "      <th>CD25</th>\n",
       "      <th>CD38</th>\n",
       "      <th>CD127</th>\n",
       "      <th>PD1</th>\n",
       "      <th>CCR6</th>\n",
       "      <th>ICOS</th>\n",
       "      <th>CD27</th>\n",
       "      <th>CCR7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>memory</th>\n",
       "      <td>22.444963</td>\n",
       "      <td>8083.472168</td>\n",
       "      <td>-67.636757</td>\n",
       "      <td>48.388863</td>\n",
       "      <td>128.510620</td>\n",
       "      <td>317.221191</td>\n",
       "      <td>153.553589</td>\n",
       "      <td>-477.221954</td>\n",
       "      <td>2538.507568</td>\n",
       "      <td>1307.797119</td>\n",
       "      <td>462.951263</td>\n",
       "      <td>260.562439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memory</th>\n",
       "      <td>46.322460</td>\n",
       "      <td>6292.529297</td>\n",
       "      <td>262.460266</td>\n",
       "      <td>341.741608</td>\n",
       "      <td>349.631439</td>\n",
       "      <td>638.302551</td>\n",
       "      <td>0.721993</td>\n",
       "      <td>1040.776855</td>\n",
       "      <td>2116.770264</td>\n",
       "      <td>6056.441895</td>\n",
       "      <td>8753.997070</td>\n",
       "      <td>1611.830200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memory</th>\n",
       "      <td>138.343292</td>\n",
       "      <td>1583.569580</td>\n",
       "      <td>535.068176</td>\n",
       "      <td>-177.763977</td>\n",
       "      <td>54.735806</td>\n",
       "      <td>396.080292</td>\n",
       "      <td>1005.266602</td>\n",
       "      <td>-21.077803</td>\n",
       "      <td>-146.690323</td>\n",
       "      <td>818.594910</td>\n",
       "      <td>5212.695312</td>\n",
       "      <td>688.128906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memory</th>\n",
       "      <td>99.121498</td>\n",
       "      <td>390.649139</td>\n",
       "      <td>-14.418692</td>\n",
       "      <td>83.374794</td>\n",
       "      <td>63.772312</td>\n",
       "      <td>260.273926</td>\n",
       "      <td>2112.503662</td>\n",
       "      <td>888.982605</td>\n",
       "      <td>1462.157104</td>\n",
       "      <td>672.073364</td>\n",
       "      <td>2724.397949</td>\n",
       "      <td>687.440430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>memory</th>\n",
       "      <td>53.704018</td>\n",
       "      <td>3739.389404</td>\n",
       "      <td>14.945232</td>\n",
       "      <td>13.854160</td>\n",
       "      <td>561.055847</td>\n",
       "      <td>-484.798462</td>\n",
       "      <td>175.297806</td>\n",
       "      <td>-1016.438843</td>\n",
       "      <td>4490.885742</td>\n",
       "      <td>-101.385414</td>\n",
       "      <td>7555.630859</td>\n",
       "      <td>981.396484</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             CXCR3         CCR4      CD45RA      HLA-DR        CD25   \n",
       "memory   22.444963  8083.472168  -67.636757   48.388863  128.510620  \\\n",
       "memory   46.322460  6292.529297  262.460266  341.741608  349.631439   \n",
       "memory  138.343292  1583.569580  535.068176 -177.763977   54.735806   \n",
       "memory   99.121498   390.649139  -14.418692   83.374794   63.772312   \n",
       "memory   53.704018  3739.389404   14.945232   13.854160  561.055847   \n",
       "\n",
       "              CD38        CD127          PD1         CCR6         ICOS   \n",
       "memory  317.221191   153.553589  -477.221954  2538.507568  1307.797119  \\\n",
       "memory  638.302551     0.721993  1040.776855  2116.770264  6056.441895   \n",
       "memory  396.080292  1005.266602   -21.077803  -146.690323   818.594910   \n",
       "memory  260.273926  2112.503662   888.982605  1462.157104   672.073364   \n",
       "memory -484.798462   175.297806 -1016.438843  4490.885742  -101.385414   \n",
       "\n",
       "               CD27         CCR7  \n",
       "memory   462.951263   260.562439  \n",
       "memory  8753.997070  1611.830200  \n",
       "memory  5212.695312   688.128906  \n",
       "memory  2724.397949   687.440430  \n",
       "memory  7555.630859   981.396484  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "facs.to_df().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'facs' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m summary_facs \u001b[39m=\u001b[39m {\u001b[39m\"\u001b[39m\u001b[39mfeature\u001b[39m\u001b[39m\"\u001b[39m : facs\u001b[39m.\u001b[39mvar_names,\n\u001b[1;32m      2\u001b[0m            \u001b[39m\"\u001b[39m\u001b[39mmin\u001b[39m\u001b[39m\"\u001b[39m : facs\u001b[39m.\u001b[39mX\u001b[39m.\u001b[39mmin(axis \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m),\n\u001b[1;32m      3\u001b[0m            \u001b[39m\"\u001b[39m\u001b[39mmax\u001b[39m\u001b[39m\"\u001b[39m : facs\u001b[39m.\u001b[39mX\u001b[39m.\u001b[39mmax(axis \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m),\n\u001b[1;32m      4\u001b[0m            \u001b[39m\"\u001b[39m\u001b[39mmean\u001b[39m\u001b[39m\"\u001b[39m : facs\u001b[39m.\u001b[39mX\u001b[39m.\u001b[39mmean(axis \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m),\n\u001b[1;32m      5\u001b[0m            \u001b[39m\"\u001b[39m\u001b[39mvar\u001b[39m\u001b[39m\"\u001b[39m : facs\u001b[39m.\u001b[39mX\u001b[39m.\u001b[39mvar(axis \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m)}\n\u001b[1;32m      7\u001b[0m df\u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(summary_facs)\n\u001b[1;32m      8\u001b[0m df\u001b[39m.\u001b[39mto_excel(\u001b[39m\"\u001b[39m\u001b[39msummary_facs.xlsx\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39m# Save summary\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'facs' is not defined"
     ]
    }
   ],
   "source": [
    "with open ('facs_feature_names.txt', 'w') as file:  \n",
    "    for name in facs.var_names:\n",
    "        file.write(name + \"\\n\") \n",
    "\n",
    "df_facs.to_excel(\"summary_facs.xlsx\") # Save summary"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparisons <a id='compare'></a>\n",
    "\n",
    "The amin of this section is to find shared proteins and genes across modalities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_features = set(facs.var_names).union(set(cytof.var_names), set(luminex.var_names), set(bulkRNA_joined[\"gene_name\"]), set(adt_names), set(scRNA_names)) \n",
    "\n",
    "with open ('all_features.txt', 'w') as file:  \n",
    "    for name in all_features:\n",
    "        file.write(name + \"\\n\") "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Shared Protein Names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare bulkRNA and ADT\n",
    "shared_features_adt_bulkRNA = set(adt_names).intersection(set(bulkRNA_joined[\"gene_name\"]))\n",
    "\n",
    "# compare Luminex, bulkRNA and ADT\n",
    "shared_features_adt_luminex = set(adt_names).intersection(set(luminex.var_names))\n",
    "shared_features_bulkRNA_luminex = set(luminex.var_names).intersection(set(bulkRNA_joined[\"gene_name\"]))\n",
    "\n",
    "# compare cyTOF bulkRNA and ADT\n",
    "shared_features_adt_cytof = set(cytof.var_names).intersection(set(adt_names)) \n",
    "shared_features_bulkRNA_cytof = set(cytof.var_names).intersection(set(bulkRNA_joined[\"gene_name\"]))\n",
    "\n",
    "# compare FACS, bulkRNA and ADT, cyTOF\n",
    "shared_features_adt_facs = set(facs.var_names).intersection(set(adt_names)) \n",
    "shared_features_bulkRNA_facs = set(facs.var_names).intersection(set(bulkRNA_joined[\"gene_name\"]))\n",
    "shared_features_cytof_facs = set(facs.var_names).intersection(set(cytof.var_names)) \n",
    "shared_features_adt_bulkRNA_facs = shared_features_adt_facs.intersection(set(bulkRNA_joined[\"gene_name\"]))\n",
    "shared_features_adt_cytof_facs = shared_features_adt_facs.intersection(set(cytof.var_names)) \n",
    "shared_features_bulkRNA_cytof_facs = shared_features_bulkRNA_facs.intersection(set(cytof.var_names)) \n",
    "shared_features_adt_bulkRNA_cytof_facs = shared_features_adt_bulkRNA.intersection(set(cytof.var_names)) \n",
    "\n",
    "# compare FACS, bulkRNA and ADT, cyTOF, csRNA\n",
    "shared_features_adt_scRNA = set(scRNA_names).intersection(set(adt_names)) \n",
    "shared_features_bulkRNA_scRNA = set(scRNA_names).intersection(set(bulkRNA_joined[\"gene_name\"]))\n",
    "shared_features_cytof_scRNA = set(scRNA_names).intersection(set(cytof.var_names)) \n",
    "shared_features_facs_scRNA = set(scRNA_names).intersection(set(facs.var_names))\n",
    "shared_features_adt_bulkRNA_scRNA = shared_features_adt_scRNA.intersection(set(bulkRNA_joined[\"gene_name\"]))\n",
    "shared_features_adt_cytof_scRNA = shared_features_adt_scRNA.intersection(set(cytof.var_names)) \n",
    "shared_features_adt_facs_scRNA = shared_features_adt_scRNA.intersection(set(facs.var_names)) \n",
    "shared_features_bulkRNA_cytof_scRNA = shared_features_bulkRNA_scRNA.intersection(set(cytof.var_names)) \n",
    "shared_features_bulkRNA_facs_scRNA = shared_features_bulkRNA_scRNA.intersection(set(facs.var_names))\n",
    "shared_features_adt_bulkRNA_cytof_scRNA = shared_features_adt_bulkRNA.intersection(set(cytof.var_names)) \n",
    "shared_features_adt_bulkRNA_facs_scRNA = shared_features_adt_bulkRNA.intersection(set(facs.var_names)) \n",
    "shared_features_adt_cytof_facs_scRNA = shared_features_adt_cytof.intersection(set(facs.var_names)) \n",
    "shared_features_bulkRNA_facs_scRNA_cytof = shared_features_bulkRNA_facs_scRNA.intersection(set(cytof.var_names)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open ('shared_ADT_bulkRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_bulkRNA:\n",
    "        file.write(name + \"\\n\")  \n",
    "\n",
    "\n",
    "with open ('shared_ADT_Luminex.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_luminex:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_bulkRNA_Luminex.txt', 'w') as file:  \n",
    "    for name in shared_features_bulkRNA_luminex:\n",
    "        file.write(name + \"\\n\")  \n",
    "\n",
    "\n",
    "with open ('shared_ADT_cyTOF.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_cytof:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_bulkRNA_cyTOF.txt', 'w') as file:  \n",
    "    for name in shared_features_bulkRNA_cytof:\n",
    "        file.write(name + \"\\n\")  \n",
    "\n",
    "\n",
    "with open ('shared_ADT_FACS.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_facs:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_bulkRNA_FACS.txt', 'w') as file:  \n",
    "    for name in shared_features_bulkRNA_facs:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_cyTOF_FACS.txt', 'w') as file:  \n",
    "    for name in shared_features_cytof_facs:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_ADT_bulkRNA_FACS.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_bulkRNA:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_ADT_cytof_FACS.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_cytof:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_bulkRNA_cytof_FACS.txt', 'w') as file:  \n",
    "    for name in shared_features_bulkRNA_cytof:\n",
    "        file.write(name + \"\\n\") \n",
    "with open ('shared_ADT_bulkRNA_cytof_FACS.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_bulkRNA_cytof_facs:\n",
    "        file.write(name + \"\\n\")       \n",
    "\n",
    "\n",
    "with open ('shared_ADT_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_scRNA:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_bulkRNA_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_bulkRNA_scRNA:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_cyTOF_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_cytof_scRNA:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_FACS_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_facs_scRNA:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_ADT_bulkRNA_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_bulkRNA:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_ADT_cytof_csRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_cytof:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_ADT_FACS_csRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_facs:\n",
    "        file.write(name + \"\\n\")  \n",
    "with open ('shared_bulkRNA_cytof_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_bulkRNA_cytof:\n",
    "        file.write(name + \"\\n\") \n",
    "with open ('shared_bulkRNA_FACS_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_bulkRNA_facs:\n",
    "        file.write(name + \"\\n\") \n",
    "with open ('shared_ADT_bulkRNA_cytof_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_bulkRNA_cytof_scRNA:\n",
    "        file.write(name + \"\\n\")       \n",
    "with open ('shared_ADT_bulkRNA_FACS_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_adt_bulkRNA_facs:\n",
    "        file.write(name + \"\\n\")    \n",
    "with open ('shared_bulkRNA_cytof_FACS_scRNA.txt', 'w') as file:  \n",
    "    for name in shared_features_bulkRNA_facs_scRNA_cytof:\n",
    "        file.write(name + \"\\n\")     "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Shared Protein Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_bulkRNA_scRNA = df_bulkRNA.set_index(\"feature\").join(df_scRNA.set_index(\"feature\"), lsuffix=\"_bulk\", rsuffix=\"_sc\", how=\"inner\")\n",
    "stats_cytof_scRNA = df_cytof.set_index(\"feature\").join(df_scRNA.set_index(\"feature\"), lsuffix=\"_cytof\", rsuffix=\"_scRNA\", how = \"inner\")\n",
    "stats_adt_scRNA = df_adt.set_index(\"feature\").join(df_scRNA.set_index(\"feature\"), lsuffix=\"_adt\", rsuffix=\"_scRNA\", how = \"inner\") \n",
    "stats_facs_scRNA = df_facs.set_index(\"feature\").join(df_scRNA.set_index(\"feature\"), lsuffix=\"_facs\", rsuffix=\"_scRNA\", how = \"inner\") \n",
    "\n",
    "stats_bulkRNA_cytof = df_bulkRNA.set_index(\"feature\").join(df_cytof.set_index(\"feature\"), lsuffix=\"_bulk\", rsuffix=\"_cytof\", how=\"inner\")\n",
    "stats_adt_cytof = df_adt.set_index(\"feature\").join(df_cytof.set_index(\"feature\"), lsuffix=\"_adt\", rsuffix=\"_cytof\", how = \"inner\") \n",
    "stats_cytof_facs = df_facs.set_index(\"feature\").join(df_cytof.set_index(\"feature\"), lsuffix=\"_facs\", rsuffix=\"_cytof\", how = \"inner\") \n",
    "\n",
    "stats_adt_bulkRNA = df_adt.set_index(\"feature\").join(df_bulkRNA.set_index(\"feature\"), lsuffix=\"_adt\", rsuffix=\"_bulkRNA\", how = \"inner\") \n",
    "stats_bulkRNA_facs = df_facs.set_index(\"feature\").join(df_bulkRNA.set_index(\"feature\"), lsuffix=\"_facs\", rsuffix=\"_bulkRNA\", how = \"inner\") \n",
    "\n",
    "stats_adt_facs = df_facs.set_index(\"feature\").join(df_adt.set_index(\"feature\"), lsuffix=\"_facs\", rsuffix=\"_adt\", how = \"inner\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_bulkRNA_scRNA.to_excel(\"stats_bulkRNA_scRNA.xlsx\") \n",
    "stats_cytof_scRNA.to_excel(\"stats_cytof_scRNA.xlsx\") \n",
    "stats_adt_scRNA.to_excel(\"stats_adt_scRNA.xlsx\") \n",
    "stats_facs_scRNA.to_excel(\"stats_facs_scRNA.xlsx\") \n",
    "\n",
    "stats_bulkRNA_cytof.to_excel(\"stats_bulkRNA_cytof.xlsx\") \n",
    "stats_adt_cytof.to_excel(\"stats_adt_cytof.xlsx\") \n",
    "stats_cytof_facs.to_excel(\"stats_cytof_facs.xlsx\") \n",
    "\n",
    "stats_adt_bulkRNA.to_excel(\"stats_adt_bulkRNA.xlsx\") \n",
    "stats_bulkRNA_facs.to_excel(\"stats_bulkRNA_facs.xlsx\") \n",
    "\n",
    "stats_adt_facs.to_excel(\"stats_adt_facs.xlsx\") "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pseudo Bulking <a id='aggregate'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Annotation_cluster_id</th>\n",
       "      <th>Annotation_cluster_name</th>\n",
       "      <th>Annotation_minor_subset</th>\n",
       "      <th>Annotation_major_subset</th>\n",
       "      <th>Annotation_cell_type</th>\n",
       "      <th>GEX_region</th>\n",
       "      <th>QC_ngenes</th>\n",
       "      <th>QC_total_UMI</th>\n",
       "      <th>QC_pct_mitochondrial</th>\n",
       "      <th>QC_scrub_doublet_scores</th>\n",
       "      <th>...</th>\n",
       "      <th>Requiredvasoactive</th>\n",
       "      <th>Respiratorysupport</th>\n",
       "      <th>SARSCoV2PCR</th>\n",
       "      <th>Outcome</th>\n",
       "      <th>TimeSinceOnset</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>DiseaseClassification</th>\n",
       "      <th>Pool_ID</th>\n",
       "      <th>Channel_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAAGTGG-1-gPlexA1</th>\n",
       "      <td>20120.0</td>\n",
       "      <td>NK.CD16hi.1</td>\n",
       "      <td>NK.CD16hi</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>B: TEM/prolif. T/NK cells</td>\n",
       "      <td>1159</td>\n",
       "      <td>2684</td>\n",
       "      <td>1.862891</td>\n",
       "      <td>0.031883</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>COVID-19;MONDO:0100096</td>\n",
       "      <td>gPlexA</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCGGATCA-1-gPlexA1</th>\n",
       "      <td>20011.0</td>\n",
       "      <td>CD8.TEMRA.1</td>\n",
       "      <td>CD8.TEMRA</td>\n",
       "      <td>CD8</td>\n",
       "      <td>T</td>\n",
       "      <td>B: TEM/prolif. T/NK cells</td>\n",
       "      <td>1348</td>\n",
       "      <td>3162</td>\n",
       "      <td>1.138520</td>\n",
       "      <td>0.041541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>COVID-19;MONDO:0100096</td>\n",
       "      <td>gPlexA</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGACATTA-1-gPlexA1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>D: B/Plasma cells</td>\n",
       "      <td>937</td>\n",
       "      <td>2579</td>\n",
       "      <td>0.891819</td>\n",
       "      <td>0.003108</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>COVID-19;MONDO:0100096</td>\n",
       "      <td>gPlexA</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCGACAT-1-gPlexA1</th>\n",
       "      <td>30400.0</td>\n",
       "      <td>ncMono</td>\n",
       "      <td>ncMono</td>\n",
       "      <td>ncMono</td>\n",
       "      <td>MNP</td>\n",
       "      <td>C: Monocytes/cDC</td>\n",
       "      <td>788</td>\n",
       "      <td>1979</td>\n",
       "      <td>4.194037</td>\n",
       "      <td>0.068193</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>COVID-19;MONDO:0100096</td>\n",
       "      <td>gPlexA</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGGAACGG-1-gPlexA1</th>\n",
       "      <td>30100.0</td>\n",
       "      <td>cMono.LGALS2.AHNAK</td>\n",
       "      <td>cMono</td>\n",
       "      <td>cMono</td>\n",
       "      <td>MNP</td>\n",
       "      <td>C: Monocytes/cDC</td>\n",
       "      <td>1344</td>\n",
       "      <td>3084</td>\n",
       "      <td>3.728923</td>\n",
       "      <td>0.036749</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>COVID-19;MONDO:0100096</td>\n",
       "      <td>gPlexA</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGGCAAAC-1-gPlexK7</th>\n",
       "      <td>30000.0</td>\n",
       "      <td>cMono.S100A8/9/12hi.HMGB2</td>\n",
       "      <td>cMono</td>\n",
       "      <td>cMono</td>\n",
       "      <td>MNP</td>\n",
       "      <td>C: Monocytes/cDC</td>\n",
       "      <td>1316</td>\n",
       "      <td>3722</td>\n",
       "      <td>3.519613</td>\n",
       "      <td>0.068154</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>NA</td>\n",
       "      <td>gPlexK</td>\n",
       "      <td>K7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTACCGAT-1-gPlexK7</th>\n",
       "      <td>20211.0</td>\n",
       "      <td>CD8.TEM</td>\n",
       "      <td>CD8.TEM</td>\n",
       "      <td>CD8</td>\n",
       "      <td>T</td>\n",
       "      <td>B: TEM/prolif. T/NK cells</td>\n",
       "      <td>1157</td>\n",
       "      <td>3318</td>\n",
       "      <td>1.476793</td>\n",
       "      <td>0.494080</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>Influenza;MONDO:0005812</td>\n",
       "      <td>gPlexK</td>\n",
       "      <td>K7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCTCTAGC-1-gPlexK7</th>\n",
       "      <td>21311.0</td>\n",
       "      <td>CD8.TEMRA.mitohi.2</td>\n",
       "      <td>CD8.mitohi</td>\n",
       "      <td>CD8</td>\n",
       "      <td>T</td>\n",
       "      <td>B: TEM/prolif. T/NK cells</td>\n",
       "      <td>502</td>\n",
       "      <td>627</td>\n",
       "      <td>7.017544</td>\n",
       "      <td>0.019165</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>COVID-19;MONDO:0100096</td>\n",
       "      <td>gPlexK</td>\n",
       "      <td>K7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGAGGTAG-1-gPlexK7</th>\n",
       "      <td>30300.0</td>\n",
       "      <td>cMono.LGALS2.PSME2.IFITM3hi</td>\n",
       "      <td>cMono</td>\n",
       "      <td>cMono</td>\n",
       "      <td>MNP</td>\n",
       "      <td>C: Monocytes/cDC</td>\n",
       "      <td>805</td>\n",
       "      <td>1612</td>\n",
       "      <td>2.233251</td>\n",
       "      <td>0.050830</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>COVID-19;MONDO:0100096</td>\n",
       "      <td>gPlexK</td>\n",
       "      <td>K7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTCCCTGA-1-gPlexK7</th>\n",
       "      <td>30100.0</td>\n",
       "      <td>cMono.LGALS2.AHNAK</td>\n",
       "      <td>cMono</td>\n",
       "      <td>cMono</td>\n",
       "      <td>MNP</td>\n",
       "      <td>C: Monocytes/cDC</td>\n",
       "      <td>625</td>\n",
       "      <td>1184</td>\n",
       "      <td>2.702703</td>\n",
       "      <td>0.025384</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Blood;UBERON:0000178</td>\n",
       "      <td>COVID-19;MONDO:0100096</td>\n",
       "      <td>gPlexK</td>\n",
       "      <td>K7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>836148 rows × 103 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Annotation_cluster_id   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1                20120.0  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1                20011.0   \n",
       "AAACCTGAGGACATTA-1-gPlexA1                    NaN   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1                30400.0   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1                30100.0   \n",
       "...                                           ...   \n",
       "TTTGTCAGTGGCAAAC-1-gPlexK7                30000.0   \n",
       "TTTGTCAGTTACCGAT-1-gPlexK7                20211.0   \n",
       "TTTGTCATCCTCTAGC-1-gPlexK7                21311.0   \n",
       "TTTGTCATCGAGGTAG-1-gPlexK7                30300.0   \n",
       "TTTGTCATCTCCCTGA-1-gPlexK7                30100.0   \n",
       "\n",
       "                                Annotation_cluster_name   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1                  NK.CD16hi.1  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1                  CD8.TEMRA.1   \n",
       "AAACCTGAGGACATTA-1-gPlexA1                          nan   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1                       ncMono   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1           cMono.LGALS2.AHNAK   \n",
       "...                                                 ...   \n",
       "TTTGTCAGTGGCAAAC-1-gPlexK7    cMono.S100A8/9/12hi.HMGB2   \n",
       "TTTGTCAGTTACCGAT-1-gPlexK7                      CD8.TEM   \n",
       "TTTGTCATCCTCTAGC-1-gPlexK7           CD8.TEMRA.mitohi.2   \n",
       "TTTGTCATCGAGGTAG-1-gPlexK7  cMono.LGALS2.PSME2.IFITM3hi   \n",
       "TTTGTCATCTCCCTGA-1-gPlexK7           cMono.LGALS2.AHNAK   \n",
       "\n",
       "                           Annotation_minor_subset Annotation_major_subset   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1               NK.CD16hi                      NK  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1               CD8.TEMRA                     CD8   \n",
       "AAACCTGAGGACATTA-1-gPlexA1                     nan                     nan   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1                  ncMono                  ncMono   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1                   cMono                   cMono   \n",
       "...                                            ...                     ...   \n",
       "TTTGTCAGTGGCAAAC-1-gPlexK7                   cMono                   cMono   \n",
       "TTTGTCAGTTACCGAT-1-gPlexK7                 CD8.TEM                     CD8   \n",
       "TTTGTCATCCTCTAGC-1-gPlexK7              CD8.mitohi                     CD8   \n",
       "TTTGTCATCGAGGTAG-1-gPlexK7                   cMono                   cMono   \n",
       "TTTGTCATCTCCCTGA-1-gPlexK7                   cMono                   cMono   \n",
       "\n",
       "                           Annotation_cell_type                 GEX_region   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1                   NK  B: TEM/prolif. T/NK cells  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1                    T  B: TEM/prolif. T/NK cells   \n",
       "AAACCTGAGGACATTA-1-gPlexA1                  nan          D: B/Plasma cells   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1                  MNP           C: Monocytes/cDC   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1                  MNP           C: Monocytes/cDC   \n",
       "...                                         ...                        ...   \n",
       "TTTGTCAGTGGCAAAC-1-gPlexK7                  MNP           C: Monocytes/cDC   \n",
       "TTTGTCAGTTACCGAT-1-gPlexK7                    T  B: TEM/prolif. T/NK cells   \n",
       "TTTGTCATCCTCTAGC-1-gPlexK7                    T  B: TEM/prolif. T/NK cells   \n",
       "TTTGTCATCGAGGTAG-1-gPlexK7                  MNP           C: Monocytes/cDC   \n",
       "TTTGTCATCTCCCTGA-1-gPlexK7                  MNP           C: Monocytes/cDC   \n",
       "\n",
       "                            QC_ngenes  QC_total_UMI  QC_pct_mitochondrial   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1       1159          2684              1.862891  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1       1348          3162              1.138520   \n",
       "AAACCTGAGGACATTA-1-gPlexA1        937          2579              0.891819   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1        788          1979              4.194037   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1       1344          3084              3.728923   \n",
       "...                               ...           ...                   ...   \n",
       "TTTGTCAGTGGCAAAC-1-gPlexK7       1316          3722              3.519613   \n",
       "TTTGTCAGTTACCGAT-1-gPlexK7       1157          3318              1.476793   \n",
       "TTTGTCATCCTCTAGC-1-gPlexK7        502           627              7.017544   \n",
       "TTTGTCATCGAGGTAG-1-gPlexK7        805          1612              2.233251   \n",
       "TTTGTCATCTCCCTGA-1-gPlexK7        625          1184              2.702703   \n",
       "\n",
       "                            QC_scrub_doublet_scores  ... Requiredvasoactive   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1                 0.031883  ...                1.0  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1                 0.041541  ...                0.0   \n",
       "AAACCTGAGGACATTA-1-gPlexA1                 0.003108  ...                NaN   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1                 0.068193  ...                1.0   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1                 0.036749  ...                0.0   \n",
       "...                                             ...  ...                ...   \n",
       "TTTGTCAGTGGCAAAC-1-gPlexK7                 0.068154  ...                NaN   \n",
       "TTTGTCAGTTACCGAT-1-gPlexK7                 0.494080  ...                1.0   \n",
       "TTTGTCATCCTCTAGC-1-gPlexK7                 0.019165  ...                NaN   \n",
       "TTTGTCATCGAGGTAG-1-gPlexK7                 0.050830  ...                0.0   \n",
       "TTTGTCATCTCCCTGA-1-gPlexK7                 0.025384  ...                0.0   \n",
       "\n",
       "                           Respiratorysupport  SARSCoV2PCR  Outcome   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1                1.0            1      2.0  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1                4.0            1      5.0   \n",
       "AAACCTGAGGACATTA-1-gPlexA1                4.0            1      6.0   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1                1.0            1      2.0   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1                3.0            1      4.0   \n",
       "...                                       ...          ...      ...   \n",
       "TTTGTCAGTGGCAAAC-1-gPlexK7                4.0            0      6.0   \n",
       "TTTGTCAGTTACCGAT-1-gPlexK7                1.0            0      2.0   \n",
       "TTTGTCATCCTCTAGC-1-gPlexK7                4.0            1      6.0   \n",
       "TTTGTCATCGAGGTAG-1-gPlexK7                5.0            1      1.0   \n",
       "TTTGTCATCTCCCTGA-1-gPlexK7                5.0            1      1.0   \n",
       "\n",
       "                            TimeSinceOnset  Ethnicity                Tissue   \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1            12.0    Unknown  Blood;UBERON:0000178  \\\n",
       "AAACCTGAGCGGATCA-1-gPlexA1            12.0    Unknown  Blood;UBERON:0000178   \n",
       "AAACCTGAGGACATTA-1-gPlexA1            17.0    Unknown  Blood;UBERON:0000178   \n",
       "AAACCTGAGGCGACAT-1-gPlexA1            14.0    Unknown  Blood;UBERON:0000178   \n",
       "AAACCTGAGGGAACGG-1-gPlexA1             6.0    Unknown  Blood;UBERON:0000178   \n",
       "...                                    ...        ...                   ...   \n",
       "TTTGTCAGTGGCAAAC-1-gPlexK7             NaN    Unknown  Blood;UBERON:0000178   \n",
       "TTTGTCAGTTACCGAT-1-gPlexK7            11.0    Unknown  Blood;UBERON:0000178   \n",
       "TTTGTCATCCTCTAGC-1-gPlexK7             7.0    Unknown  Blood;UBERON:0000178   \n",
       "TTTGTCATCGAGGTAG-1-gPlexK7             3.0    Unknown  Blood;UBERON:0000178   \n",
       "TTTGTCATCTCCCTGA-1-gPlexK7             3.0    Unknown  Blood;UBERON:0000178   \n",
       "\n",
       "                              DiseaseClassification Pool_ID Channel_ID  \n",
       "AAACCTGAGAAAGTGG-1-gPlexA1   COVID-19;MONDO:0100096  gPlexA         A1  \n",
       "AAACCTGAGCGGATCA-1-gPlexA1   COVID-19;MONDO:0100096  gPlexA         A1  \n",
       "AAACCTGAGGACATTA-1-gPlexA1   COVID-19;MONDO:0100096  gPlexA         A1  \n",
       "AAACCTGAGGCGACAT-1-gPlexA1   COVID-19;MONDO:0100096  gPlexA         A1  \n",
       "AAACCTGAGGGAACGG-1-gPlexA1   COVID-19;MONDO:0100096  gPlexA         A1  \n",
       "...                                             ...     ...        ...  \n",
       "TTTGTCAGTGGCAAAC-1-gPlexK7                       NA  gPlexK         K7  \n",
       "TTTGTCAGTTACCGAT-1-gPlexK7  Influenza;MONDO:0005812  gPlexK         K7  \n",
       "TTTGTCATCCTCTAGC-1-gPlexK7   COVID-19;MONDO:0100096  gPlexK         K7  \n",
       "TTTGTCATCGAGGTAG-1-gPlexK7   COVID-19;MONDO:0100096  gPlexK         K7  \n",
       "TTTGTCATCTCCCTGA-1-gPlexK7   COVID-19;MONDO:0100096  gPlexK         K7  \n",
       "\n",
       "[836148 rows x 103 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "facs.to_df().groupby(facs.obs.sample_id).sum().to_excel(\"pseudo_bulk_facs.xlsx\")\n",
    "cytof.to_df().groupby(by = [cytof.obs.major_cell_type, cytof.obs.patient_id]).sum().to_excel(\"pseudo_bulk_cytof.xlsx\")\n",
    "\n",
    "\n",
    "#scRNA.obs\n",
    "#ADT"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
